<?xml version="1.0" encoding="UTF-8"?><rss version="2.0"
	xmlns:content="http://purl.org/rss/1.0/modules/content/"
	xmlns:wfw="http://wellformedweb.org/CommentAPI/"
	xmlns:dc="http://purl.org/dc/elements/1.1/"
	xmlns:atom="http://www.w3.org/2005/Atom"
	xmlns:sy="http://purl.org/rss/1.0/modules/syndication/"
	xmlns:slash="http://purl.org/rss/1.0/modules/slash/"
	>

<channel>
	<title>Sam &#38; Max &#187; exceptions</title>
	<atom:link href="http://sametmax.com/tag/exceptions/feed/" rel="self" type="application/rss+xml" />
	<link>http://sametmax.com</link>
	<description>Du code, du cul</description>
	<lastBuildDate>Sat, 07 Nov 2015 10:56:13 +0000</lastBuildDate>
	<language>en-US</language>
	<sy:updatePeriod>hourly</sy:updatePeriod>
	<sy:updateFrequency>1</sy:updateFrequency>
	<generator>http://wordpress.org/?v=4.1</generator>
	<item>
		<title>Gestion des erreurs en Python 12</title>
		<link>http://sametmax.com/gestion-des-erreurs-en-python/</link>
		<comments>http://sametmax.com/gestion-des-erreurs-en-python/#comments</comments>
		<pubDate>Tue, 20 Jan 2015 17:46:12 +0000</pubDate>
		<dc:creator><![CDATA[Sam]]></dc:creator>
				<category><![CDATA[Programmation]]></category>
		<category><![CDATA[else]]></category>
		<category><![CDATA[erreurs]]></category>
		<category><![CDATA[except]]></category>
		<category><![CDATA[exceptions]]></category>
		<category><![CDATA[finally]]></category>
		<category><![CDATA[python]]></category>
		<category><![CDATA[try]]></category>

		<guid isPermaLink="false">http://sametmax.com/?p=15786</guid>
		<description><![CDATA[Plus on code, plus on oublie ce que c&#8217;était quand on a débuté. Même moi, et pourtant je fais un gros effort pour essayer de me replonger dans cet état d&#8217;esprit. Dernièrement j&#8217;ai eu plusieurs interactions qui m&#8217;ont mis un petit taquet derrière la tête : - &#8220;Mec, les exceptions, y a plein de personnes [&#8230;]]]></description>
				<content:encoded><![CDATA[<p>Plus on code, plus on oublie ce que c&#8217;était quand on a débuté. Même moi, et pourtant je fais un gros effort pour essayer de me replonger dans cet état d&#8217;esprit.</p>
<p>Dernièrement j&#8217;ai eu plusieurs interactions qui m&#8217;ont mis un petit taquet derrière la tête :</p>
<blockquote><p>- &#8220;Mec, les exceptions, y a plein de personnes qui pigent pas.&#8221;</p>
<p>&#8211; &#8220;Nannnn, mais quand même, les exceptions&#8230;&#8221;</p>
<p>&#8211; &#8220;Si, les exceptions.&#8221;</p>
<p>&#8211; &#8220;Nan, vraiment ?&#8221;</p>
<p>&#8211; &#8220;Oui, vraiment.&#8221;</p></blockquote>
<p>Donc, back to black, et petit tour de la gestion d&#8217;erreur en Python.</p>
<p>Et de la musique.</p>
<p><span class='embed-youtube' style='text-align:center; display: block;'><iframe class='youtube-player' type='text/html' width='1170' height='689' src='http://www.youtube.com/embed/8inJtTG_DuU?version=3&#038;rel=1&#038;fs=1&#038;showsearch=0&#038;showinfo=1&#038;iv_load_policy=1&#038;wmode=transparent' frameborder='0' allowfullscreen='true'></iframe></span></p>
<h2>La notion d&#8217;exception</h2>
<p>Une exception est un mécanisme d&#8217;interruption du programme utilisé pour signaler que quelque chose d&#8217;anormal est en train de se produire.</p>
<p>On les rencontre dans de nombreux cas, mais souvent, c&#8217;est dans le cadre d&#8217;erreurs. Par exemple, Python va lever une exception dans les cas suivant :</p>

<div class="wp_syntax"><table><tr><td class="code"><pre class="python" style="font-family:monospace;"><span style="color: #66cc66;">&gt;&gt;&gt;</span> <span style="color: #ff4500;">1</span> / <span style="color: #ff4500;">0</span> <span style="color: #808080; font-style: italic;"># division par zéro</span>
&nbsp;
Traceback <span style="color: black;">&#40;</span>most recent call last<span style="color: black;">&#41;</span>:
  File <span style="color: #483d8b;">&quot;&lt;ipython-input-2-b710d87c980c&gt;&quot;</span><span style="color: #66cc66;">,</span> line <span style="color: #ff4500;">1</span><span style="color: #66cc66;">,</span> <span style="color: #ff7700;font-weight:bold;">in</span> <span style="color: #66cc66;">&lt;</span>module<span style="color: #66cc66;">&gt;</span>
    <span style="color: #ff4500;">1</span> / <span style="color: #ff4500;">0</span>
<span style="color: #008000;">ZeroDivisionError</span>: integer division <span style="color: #ff7700;font-weight:bold;">or</span> modulo by zero
&nbsp;
<span style="color: #66cc66;">&gt;&gt;&gt;</span> l <span style="color: #66cc66;">=</span> <span style="color: black;">&#91;</span><span style="color: #ff4500;">1</span><span style="color: #66cc66;">,</span> <span style="color: #ff4500;">2</span><span style="color: #66cc66;">,</span> <span style="color: #ff4500;">3</span><span style="color: black;">&#93;</span>
<span style="color: #66cc66;">&gt;&gt;&gt;</span> l<span style="color: black;">&#91;</span><span style="color: #ff4500;">100</span><span style="color: black;">&#93;</span> <span style="color: #808080; font-style: italic;"># dépassement d'un tableau</span>
&nbsp;
Traceback <span style="color: black;">&#40;</span>most recent call last<span style="color: black;">&#41;</span>:
  File <span style="color: #483d8b;">&quot;&lt;ipython-input-4-3e7ce3111e95&gt;&quot;</span><span style="color: #66cc66;">,</span> line <span style="color: #ff4500;">1</span><span style="color: #66cc66;">,</span> <span style="color: #ff7700;font-weight:bold;">in</span> <span style="color: #66cc66;">&lt;</span>module<span style="color: #66cc66;">&gt;</span>
    l<span style="color: black;">&#91;</span><span style="color: #ff4500;">100</span><span style="color: black;">&#93;</span>
<span style="color: #008000;">IndexError</span>: <span style="color: #008000;">list</span> index out of <span style="color: #008000;">range</span>
&nbsp;
<span style="color: #66cc66;">&gt;&gt;&gt;</span> d <span style="color: #66cc66;">=</span> <span style="color: black;">&#123;</span><span style="color: #483d8b;">'cle'</span>: <span style="color: #483d8b;">'valeur'</span><span style="color: black;">&#125;</span>
<span style="color: #66cc66;">&gt;&gt;&gt;</span> d<span style="color: black;">&#91;</span><span style="color: #483d8b;">'nope'</span><span style="color: black;">&#93;</span> <span style="color: #808080; font-style: italic;"># clé de dico inconnue</span>
&nbsp;
Traceback <span style="color: black;">&#40;</span>most recent call last<span style="color: black;">&#41;</span>:
  File <span style="color: #483d8b;">&quot;&lt;ipython-input-6-9d965227e4b0&gt;&quot;</span><span style="color: #66cc66;">,</span> line <span style="color: #ff4500;">1</span><span style="color: #66cc66;">,</span> <span style="color: #ff7700;font-weight:bold;">in</span> <span style="color: #66cc66;">&lt;</span>module<span style="color: #66cc66;">&gt;</span>
    d<span style="color: black;">&#91;</span><span style="color: #483d8b;">'nope'</span><span style="color: black;">&#93;</span>
<span style="color: #008000;">KeyError</span>: <span style="color: #483d8b;">'nope'</span>
&nbsp;
<span style="color: #66cc66;">&gt;&gt;&gt;</span> <span style="color: #ff4500;">1</span> + <span style="color: #483d8b;">&quot;banane&quot;</span> <span style="color: #808080; font-style: italic;"># opération entre types incompatibles</span>
&nbsp;
Traceback <span style="color: black;">&#40;</span>most recent call last<span style="color: black;">&#41;</span>:
  File <span style="color: #483d8b;">&quot;&lt;ipython-input-7-5147635e41b3&gt;&quot;</span><span style="color: #66cc66;">,</span> line <span style="color: #ff4500;">1</span><span style="color: #66cc66;">,</span> <span style="color: #ff7700;font-weight:bold;">in</span> <span style="color: #66cc66;">&lt;</span>module<span style="color: #66cc66;">&gt;</span>
    <span style="color: #ff4500;">1</span> + <span style="color: #483d8b;">&quot;banane&quot;</span>
<span style="color: #008000;">TypeError</span>: unsupported operand <span style="color: #008000;">type</span><span style="color: black;">&#40;</span>s<span style="color: black;">&#41;</span> <span style="color: #ff7700;font-weight:bold;">for</span> +: <span style="color: #483d8b;">'int'</span> <span style="color: #ff7700;font-weight:bold;">and</span> <span style="color: #483d8b;">'str'</span>
&nbsp;
<span style="color: #66cc66;">&gt;&gt;&gt;</span> <span style="color: #ff7700;font-weight:bold;">import</span> nawak <span style="color: #808080; font-style: italic;"># ce module n'existe pas</span>
&nbsp;
Traceback <span style="color: black;">&#40;</span>most recent call last<span style="color: black;">&#41;</span>:
  File <span style="color: #483d8b;">&quot;&lt;ipython-input-12-c948b8798146&gt;&quot;</span><span style="color: #66cc66;">,</span> line <span style="color: #ff4500;">1</span><span style="color: #66cc66;">,</span> <span style="color: #ff7700;font-weight:bold;">in</span> <span style="color: #66cc66;">&lt;</span>module<span style="color: #66cc66;">&gt;</span>
    <span style="color: #ff7700;font-weight:bold;">import</span> nawak
<span style="color: #008000;">ImportError</span>: No module named nawak
&nbsp;
<span style="color: #66cc66;">&gt;&gt;&gt;</span> <span style="color: #008000;">open</span><span style="color: black;">&#40;</span><span style="color: #483d8b;">'nawak'</span><span style="color: black;">&#41;</span> <span style="color: #808080; font-style: italic;"># le fichier n'existe pas</span>
&nbsp;
Traceback <span style="color: black;">&#40;</span>most recent call last<span style="color: black;">&#41;</span>:
  File <span style="color: #483d8b;">&quot;&lt;ipython-input-13-4d15cfd2ffd1&gt;&quot;</span><span style="color: #66cc66;">,</span> line <span style="color: #ff4500;">1</span><span style="color: #66cc66;">,</span> <span style="color: #ff7700;font-weight:bold;">in</span> <span style="color: #66cc66;">&lt;</span>module<span style="color: #66cc66;">&gt;</span>
    <span style="color: #008000;">open</span><span style="color: black;">&#40;</span><span style="color: #483d8b;">'nawak'</span><span style="color: black;">&#41;</span>
<span style="color: #008000;">IOError</span>: <span style="color: black;">&#91;</span>Errno <span style="color: #ff4500;">2</span><span style="color: black;">&#93;</span> No such <span style="color: #008000;">file</span> <span style="color: #ff7700;font-weight:bold;">or</span> directory: <span style="color: #483d8b;">'nawak'</span>
&nbsp;
<span style="color: #66cc66;">&gt;&gt;&gt;</span> <span style="color: #ff7700;font-weight:bold;">print</span><span style="color: black;">&#40;</span>nawak<span style="color: black;">&#41;</span> <span style="color: #808080; font-style: italic;"># la variable n'existe pas</span>
Traceback <span style="color: black;">&#40;</span>most recent call last<span style="color: black;">&#41;</span>:
  File <span style="color: #483d8b;">&quot;&lt;ipython-input-15-f47a27cedc12&gt;&quot;</span><span style="color: #66cc66;">,</span> line <span style="color: #ff4500;">1</span><span style="color: #66cc66;">,</span> <span style="color: #ff7700;font-weight:bold;">in</span> <span style="color: #66cc66;">&lt;</span>module<span style="color: #66cc66;">&gt;</span>
    <span style="color: #ff7700;font-weight:bold;">print</span><span style="color: black;">&#40;</span>nawak<span style="color: black;">&#41;</span>
<span style="color: #008000;">NameError</span>: name <span style="color: #483d8b;">'nawak'</span> <span style="color: #ff7700;font-weight:bold;">is</span> <span style="color: #ff7700;font-weight:bold;">not</span> defined
&nbsp;
<span style="color: #66cc66;">&gt;&gt;&gt;</span> n a w a k <span style="color: #808080; font-style: italic;"># erreur de syntaxe</span>
  File <span style="color: #483d8b;">&quot;&lt;ipython-input-16-532bf025442e&gt;&quot;</span><span style="color: #66cc66;">,</span> line <span style="color: #ff4500;">1</span>
    n a w a k <span style="color: #808080; font-style: italic;"># erreur de syntaxe</span>
      ^
<span style="color: #008000;">SyntaxError</span>: invalid syntax
&nbsp;
<span style="color: #66cc66;">&gt;&gt;&gt;</span> <span style="color: #008000;">int</span><span style="color: black;">&#40;</span><span style="color: #483d8b;">'a'</span><span style="color: black;">&#41;</span> <span style="color: #808080; font-style: italic;"># la valeur passée n'a aucun sens</span>
&nbsp;
Traceback <span style="color: black;">&#40;</span>most recent call last<span style="color: black;">&#41;</span>:
  File <span style="color: #483d8b;">&quot;&lt;ipython-input-17-b3c3f4515dd4&gt;&quot;</span><span style="color: #66cc66;">,</span> line <span style="color: #ff4500;">1</span><span style="color: #66cc66;">,</span> <span style="color: #ff7700;font-weight:bold;">in</span> <span style="color: #66cc66;">&lt;</span>module<span style="color: #66cc66;">&gt;</span>
    <span style="color: #008000;">int</span><span style="color: black;">&#40;</span><span style="color: #483d8b;">'a'</span><span style="color: black;">&#41;</span>
<span style="color: #008000;">ValueError</span>: invalid literal <span style="color: #ff7700;font-weight:bold;">for</span> <span style="color: #008000;">int</span><span style="color: black;">&#40;</span><span style="color: black;">&#41;</span> <span style="color: #ff7700;font-weight:bold;">with</span> base <span style="color: #ff4500;">10</span>: <span style="color: #483d8b;">'a'</span></pre></td></tr></table></div>

<p>De base, il y a un paquet d&#8217;exceptions en Python, et celles fournies par défaut sont listées <a href="https://docs.python.org/2/library/exceptions.html#exceptions.TypeError">ici</a>.</p>
<p>Malgré l&#8217;abondance de ces exceptions, vous pouvez remarquer un motif récurrent quand Python crash et affiche l&#8217;erreur :</p>
<pre>Gros pâté de texte.
Gros pâté de texte.
Gros pâté de texte.
NomDeLExceptionError: description de l'erreur.</pre>
<p>En effet, non seulement une exception interrompt le programme, mais elle collecte des informations sur la source de l&#8217;erreur afin qu&#8217;au moment où ça crash, le développeur ait de quoi déboguer.</p>
<p>Une exception va donc normalement déclencher un affichage à la fin avec toutes ces infos, et sur la dernière ligne, le type de l&#8217;exception (ValueError, IndexError, etc) ainsi qu&#8217;une description de ce qui a causé l&#8217;erreur.</p>
<p>Néanmoins, le plus intéressant est ce qu&#8217;il y a au dessus : le gros pâté de texte. C&#8217;est ce qu&#8217;on appelle une stack trace, et ça représente la pile d&#8217;appels qui ont amené à cette erreur. Chaque ligne de la stack trace va vous donner le chemin d&#8217;un fichier de code, et une ligne.</p>
<p>Une erreur se lit donc à l&#8217;envers, de bas en haut.</p>
<p>Vous lisez d&#8217;abord le nom de l&#8217;erreur et sa cause, puis, vous remontez la stack trace ligne à ligne, pour essayer de trouver quelle ligne de quel fichier de code vous devez déboguer.</p>
<p>Par exemple, si vous avez un fichier <code>wololo.py</code> :</p>

<div class="wp_syntax"><table><tr><td class="code"><pre class="python" style="font-family:monospace;"><span style="color: #ff7700;font-weight:bold;">def</span> une_fonction<span style="color: black;">&#40;</span><span style="color: black;">&#41;</span>:
    <span style="color: #ff7700;font-weight:bold;">return</span> <span style="color: #ff4500;">1</span> / <span style="color: #ff4500;">0</span>
&nbsp;
<span style="color: #ff7700;font-weight:bold;">def</span> une_autre_fonction<span style="color: black;">&#40;</span><span style="color: black;">&#41;</span>:
    une_fonction<span style="color: black;">&#40;</span><span style="color: black;">&#41;</span>
&nbsp;
une_autre_fonction<span style="color: black;">&#40;</span><span style="color: black;">&#41;</span></pre></td></tr></table></div>

<p>Lancer ce script va vous pondre :</p>
<pre>
Traceback (most recent call last):
  File "wololo.py", line 7, in <module>
    une_autre_fonction()
  File "wololo.py", line 5, in une_autre_fonction
    une_fonction()
  File "wololo.py", line 2, in une_fonction
    return 1 / 0
ZeroDivisionError: integer division or modulo by zero
</pre>
<p>On le lit de bas en haut :</p>
<pre>ZeroDivisionError: integer division or modulo by zero</pre>
<p>Ok, donc c&#8217;est une erreur de division par 0.</p>
<pre>  File "wololo.py", line 2, in une_fonction
    return 1 / 0</pre>
<p>Elle a lieu dans mon fichier <code>wololo.py</code> à la ligne 2.</p>
<p>Mais visiblement, ce qui déclenche tout le bordel, c&#8217;est que la fonction contenant cette ligne est appelé à la ligne 5 :</p>
<pre>
  File "wololo.py", line 5, in une_autre_fonction
    une_fonction()</pre>
<p>Qui est elle même appelée à la ligne 7 :</p>
<pre>
  File "wololo.py", line 7, in <module>
    une_autre_fonction()
</pre>
<p>Avec ces infos, on peut prendre une décision : aller ligne 2 pour corriger un bug, ou aller ligne 5, ou 7 pour retirer l&#8217;appel qui cause le déclenchement du code fautif.</p>
<p>Dans ce cas, vous pourriez vous dire qu&#8217;il faut retirer la division par 0, et que les premières ligne de la stack trace ne servent à rien.</p>
<p>Mais imaginez que je change mon code :</p>

<div class="wp_syntax"><table><tr><td class="code"><pre class="python" style="font-family:monospace;"><span style="color: #ff7700;font-weight:bold;">def</span> une_fonction<span style="color: black;">&#40;</span>diviseur<span style="color: black;">&#41;</span>:
    <span style="color: #ff7700;font-weight:bold;">return</span> <span style="color: #ff4500;">1</span> / diviseur
&nbsp;
<span style="color: #ff7700;font-weight:bold;">def</span> une_autre_fonction<span style="color: black;">&#40;</span><span style="color: black;">&#41;</span>:
    une_fonction<span style="color: black;">&#40;</span>diviseur<span style="color: #66cc66;">=</span><span style="color: #ff4500;">0</span><span style="color: black;">&#41;</span>
&nbsp;
une_autre_fonction<span style="color: black;">&#40;</span><span style="color: black;">&#41;</span></pre></td></tr></table></div>

<p>A ce moment là, mon stack trace sera :</p>
<pre>Traceback (most recent call last):
  File "wololo.py", line 7, in <module>
    une_autre_fonction()
  File "wololo.py", line 5, in une_autre_fonction
    une_fonction(diviseur=0)
  File "wololo.py", line 2, in une_fonction
    return 1 / diviseur
ZeroDivisionError: integer division or modulo by zero</pre>
<p>Et là, on ne veut pas changer la ligne 2, on veut changer la ligne 5, car c&#8217;est ce paramètre qui fait qu&#8217;à la ligne 2 on a une division par 0.</p>
<p>La stack trace permet donc de remonter la chaîne de causalité de vos erreurs.</p>
<h2>Lever ses propres exceptions</h2>
<p>Quand une exception est activée, on dit qu&#8217;elle est &#8220;levée&#8221;.</p>
<p>Python va lever des exceptions de lui-même, par exemple en cas d&#8217;erreurs, mais on peut aussi lever soi-même des exceptions en faisant <code>raise NomDeLException("Message")</code>.</p>
<p>Par exemple, vous avez une fonction qui n&#8217;accepte en paramètre que comme valeur 1, 2 ou 3. Vous savez que si un utilisateur rentre quelque chose d&#8217;autre, ça va tout faire merder.</p>
<p>Vous pouvez faire ceci :</p>

<div class="wp_syntax"><table><tr><td class="code"><pre class="python" style="font-family:monospace;"><span style="color: #ff7700;font-weight:bold;">def</span> votre_super_fonction<span style="color: black;">&#40;</span>param<span style="color: black;">&#41;</span>:
    <span style="color: #ff7700;font-weight:bold;">if</span> param <span style="color: #ff7700;font-weight:bold;">not</span> <span style="color: #ff7700;font-weight:bold;">in</span> <span style="color: black;">&#40;</span><span style="color: #ff4500;">1</span><span style="color: #66cc66;">,</span> <span style="color: #ff4500;">2</span><span style="color: #66cc66;">,</span> <span style="color: #ff4500;">3</span><span style="color: black;">&#41;</span>:
        <span style="color: #ff7700;font-weight:bold;">raise</span> <span style="color: #008000;">ValueError</span><span style="color: black;">&#40;</span><span style="color: #483d8b;">&quot;'param' can only be either 1, 2 or 3&quot;</span><span style="color: black;">&#41;</span>
    <span style="color: #808080; font-style: italic;"># reste du code</span></pre></td></tr></table></div>

<p>Ainsi, si quelqu&#8217;un passe un mauvais paramètre, il est immédiatement alerté :</p>

<div class="wp_syntax"><table><tr><td class="code"><pre class="python" style="font-family:monospace;"><span style="color: #66cc66;">&gt;&gt;&gt;</span> votre_super_fonction<span style="color: black;">&#40;</span><span style="color: #ff4500;">4</span><span style="color: black;">&#41;</span>
Traceback <span style="color: black;">&#40;</span>most recent call last<span style="color: black;">&#41;</span>:
  File <span style="color: #483d8b;">&quot;&lt;ipython-input-3-bcd6e8653c83&gt;&quot;</span><span style="color: #66cc66;">,</span> line <span style="color: #ff4500;">1</span><span style="color: #66cc66;">,</span> <span style="color: #ff7700;font-weight:bold;">in</span> <span style="color: #66cc66;">&lt;</span>module<span style="color: #66cc66;">&gt;</span>
    votre_super_fonction<span style="color: black;">&#40;</span><span style="color: #ff4500;">4</span><span style="color: black;">&#41;</span>
  File <span style="color: #483d8b;">&quot;&lt;ipython-input-2-46fc7cd18c42&gt;&quot;</span><span style="color: #66cc66;">,</span> line <span style="color: #ff4500;">3</span><span style="color: #66cc66;">,</span> <span style="color: #ff7700;font-weight:bold;">in</span> votre_super_fonction
    <span style="color: #ff7700;font-weight:bold;">raise</span> <span style="color: #008000;">ValueError</span><span style="color: black;">&#40;</span><span style="color: #483d8b;">&quot;'param' can only be either 1, 2 or 3&quot;</span><span style="color: black;">&#41;</span>
<span style="color: #008000;">ValueError</span>: <span style="color: #483d8b;">'param'</span> can only be either <span style="color: #ff4500;">1</span><span style="color: #66cc66;">,</span> <span style="color: #ff4500;">2</span> <span style="color: #ff7700;font-weight:bold;">or</span> <span style="color: #ff4500;">3</span></pre></td></tr></table></div>

<p>Attention : le message ne peut pas contenir de caractères non-ASCII. Oubliez les accents.</p>
<h2>Attraper une exception</h2>
<p>Les exceptions ne sont pas un simple mécanisme de debuggage. Elles servent d&#8217;abord et avant tout à gérer les cas exceptionnels, et on peut donc les détecter, et réagir quand elles surviennent, à l&#8217;aide de l&#8217;instruction <code>try/except</code>.</p>
<p>Cela s&#8217;utilise en enrobant les lignes qui peuvent lever une exception :</p>

<div class="wp_syntax"><table><tr><td class="code"><pre class="python" style="font-family:monospace;">    <span style="color: #ff7700;font-weight:bold;">try</span>:
        <span style="color: #808080; font-style: italic;"># lignes qui peuvent</span>
        <span style="color: #808080; font-style: italic;"># lever une exception</span>
    <span style="color: #ff7700;font-weight:bold;">except</span> NomException:
        <span style="color: #808080; font-style: italic;"># faire un truc si l'exception se déclenche</span></pre></td></tr></table></div>

<p>Quand vous attrapez une exception, le programme ne plante pas. A la place, le bloc <code>except</code> correspondant au nom de l&#8217;exception est appelée.</p>
<p>Bien entendu, si une exception qui ne porte pas ce nom est levée, le programme plante.</p>
<p>Par exemple :</p>

<div class="wp_syntax"><table><tr><td class="code"><pre class="python" style="font-family:monospace;">    <span style="color: #808080; font-style: italic;"># i est définit plus haut</span>
    personnages <span style="color: #66cc66;">=</span> <span style="color: black;">&#91;</span><span style="color: #483d8b;">'iron'</span><span style="color: #66cc66;">,</span> <span style="color: #483d8b;">'super'</span><span style="color: #66cc66;">,</span> <span style="color: #483d8b;">'bat'</span><span style="color: #66cc66;">,</span> <span style="color: #483d8b;">'clepto'</span><span style="color: black;">&#93;</span>
    <span style="color: #ff7700;font-weight:bold;">try</span>:
        resultat <span style="color: #66cc66;">=</span> personnages<span style="color: black;">&#91;</span>i<span style="color: black;">&#93;</span>
    <span style="color: #808080; font-style: italic;"># i est plus grand que la taille du tableau</span>
    <span style="color: #ff7700;font-weight:bold;">except</span> <span style="color: #008000;">IndexError</span>:
        resultat <span style="color: #66cc66;">=</span> <span style="color: #008000;">None</span></pre></td></tr></table></div>

<p>Si <code>i</code> est plus grand que la taille de la liste, Python va lever une <code>IndexError</code>, qui sera attrapée. Le programme ne plantera pas, et <code>resultat</code> sera égal à <code>None</code>.</p>
<p>Si <code>i</code> est de type string, Python va lever une <code>TypeError</code>, et le programme va planter.</p>
<p>Cela suppose de savoir le nom de l&#8217;erreur qu&#8217;on veut gérer. Je recommande donc de tester le cas foireux dans le shell auparavant.</p>
<p>On peut gérer plusieurs exceptions avec un seul bloc :</p>

<div class="wp_syntax"><table><tr><td class="code"><pre class="python" style="font-family:monospace;">    <span style="color: #ff7700;font-weight:bold;">try</span>:
        <span style="color: #ff7700;font-weight:bold;">return</span> personnages<span style="color: black;">&#91;</span><span style="color: #ff4500;">100</span> / i<span style="color: black;">&#93;</span>
    <span style="color: #808080; font-style: italic;"># est exécuté pour ces deux exceptions</span>
    <span style="color: #ff7700;font-weight:bold;">except</span> <span style="color: black;">&#40;</span><span style="color: #008000;">IndexError</span><span style="color: #66cc66;">,</span> <span style="color: #008000;">ZeroDivisionError</span><span style="color: black;">&#41;</span>:
        <span style="color: #ff7700;font-weight:bold;">return</span> <span style="color: #008000;">None</span>
    <span style="color: #808080; font-style: italic;"># est exécuté pour cette exception</span>
    <span style="color: #ff7700;font-weight:bold;">except</span> <span style="color: #008000;">KeyError</span>:
        <span style="color: #ff7700;font-weight:bold;">print</span><span style="color: black;">&#40;</span><span style="color: #483d8b;">&quot;Qui est le bâtard qui a remplacé ma &quot;</span>
              <span style="color: #483d8b;">&quot;liste par un dico dans mon dos ?&quot;</span><span style="color: black;">&#41;</span></pre></td></tr></table></div>

<h2>Hiérarchie des exceptions</h2>
<p>Les exceptions sont des classes. Et si vous avez lu notre <a href="http://sametmax.com/le-guide-ultime-et-definitif-sur-la-programmation-orientee-objet-en-python-a-lusage-des-debutants-qui-sont-rassures-par-les-textes-detailles-qui-prennent-le-temps-de-tout-expliquer-partie-1/">dossier sur la POO</a>, vous savez que les classes peuvent hériter d&#8217;autres classes.</p>
<p>C&#8217;est le cas des exceptions. Celles intégrées à Python suivent la hiérarchie suivante :</p>
<pre>
BaseException
 +-- SystemExit
 +-- KeyboardInterrupt
 +-- GeneratorExit
 +-- Exception
      +-- StopIteration
      +-- ArithmeticError
      |    +-- FloatingPointError
      |    +-- OverflowError
      |    +-- ZeroDivisionError
      +-- AssertionError
      +-- AttributeError
      +-- BufferError
      +-- EnvironmentError
      |    +-- IOError
      |    +-- OSError
      |         +-- WindowsError (Windows)
      |         +-- VMSError (VMS)
      +-- EOFError
      +-- ImportError
      +-- LookupError
      |    +-- IndexError
      |    +-- KeyError
      +-- MemoryError
      +-- NameError
      |    +-- UnboundLocalError
      +-- ReferenceError
      +-- RuntimeError
      |    +-- NotImplementedError
      +-- SyntaxError
      |    +-- IndentationError
      |         +-- TabError
      +-- SystemError
      +-- TypeError
      +-- ValueError
      |    +-- UnicodeError
      |         +-- UnicodeDecodeError
      |         +-- UnicodeEncodeError
      |         +-- UnicodeTranslateError
      +-- Warning
           +-- DeprecationWarning
           +-- PendingDeprecationWarning
           +-- RuntimeWarning
           +-- SyntaxWarning
           +-- UserWarning
           +-- FutureWarning
           +-- ImportWarning
           +-- UnicodeWarning
           +-- BytesWarning
</pre>
<p>Par exemple, <code>IndexError</code> et <code>KeyError</code> héritent toutes les deux de <code>LookupError</code>.</p>
<p>Donc si vous faites :</p>

<div class="wp_syntax"><table><tr><td class="code"><pre class="python" style="font-family:monospace;">    <span style="color: #ff7700;font-weight:bold;">try</span>:
        <span style="color: #ff7700;font-weight:bold;">return</span> personnages<span style="color: black;">&#91;</span>i<span style="color: black;">&#93;</span>
    <span style="color: #ff7700;font-weight:bold;">except</span> <span style="color: #008000;">LookupError</span>:
        <span style="color: #ff7700;font-weight:bold;">return</span> <span style="color: #008000;">None</span></pre></td></tr></table></div>

<p>Vous attrapez à la fois les exceptions de type <code>IndexError</code> et <code>KeyError</code>.</p>
<p><code>LookupError</code> hérite de <code>Exception</code>, dont hérite la plupart des exceptions en Python. Donc si vous faites :</p>

<div class="wp_syntax"><table><tr><td class="code"><pre class="python" style="font-family:monospace;">    <span style="color: #ff7700;font-weight:bold;">try</span>:
        <span style="color: #ff7700;font-weight:bold;">return</span> personnages<span style="color: black;">&#91;</span><span style="color: #ff4500;">100</span> / i<span style="color: black;">&#93;</span>
    <span style="color: #ff7700;font-weight:bold;">except</span> <span style="color: #008000;">Exception</span>:
        <span style="color: #ff7700;font-weight:bold;">return</span> <span style="color: #008000;">None</span></pre></td></tr></table></div>

<p>Vous allez attraper <code>LookupError</code>, <code>IndexError</code>, <code>ZeroDivisionError</code> et probablement tout un tas de trucs auxquels vous n&#8217;avez pas pensé ce qui peut potentiellement un jour masquer un bug dans votre programme.</p>
<p>Donc <strong>gérez toujours l&#8217;erreur au plus proche de la logique de votre programme</strong>.</p>
<p>Il est d&#8217;ailleurs possible de dire à Python &#8220;tu m&#8217;attrapes tout&#8221; en omettant le nom de l&#8217;exception :</p>

<div class="wp_syntax"><table><tr><td class="code"><pre class="python" style="font-family:monospace;">    <span style="color: #ff7700;font-weight:bold;">try</span>:
        <span style="color: #ff7700;font-weight:bold;">return</span> personnages<span style="color: black;">&#91;</span><span style="color: #ff4500;">100</span> / i<span style="color: black;">&#93;</span>
    <span style="color: #ff7700;font-weight:bold;">except</span>:
        <span style="color: #ff7700;font-weight:bold;">return</span> <span style="color: #008000;">None</span></pre></td></tr></table></div>

<p>C&#8217;est généralement une mauvaise idée car si il arrive une erreur que vous n&#8217;avez pas prévue, la close <code>except</code> s&#8217;activera, masquant l&#8217;erreur, et ne vous donnera jamais la chance de découvrir le problème.</p>
<p>Sachez également que vous n&#8217;êtes pas limité à la hiérarchie offerte par Python. Vous pouvez créer vos propres exceptions en héritant d&#8217;une des exceptions existantes de Python :</p>

<div class="wp_syntax"><table><tr><td class="code"><pre class="python" style="font-family:monospace;"><span style="color: #ff7700;font-weight:bold;">class</span> MonProjetError<span style="color: black;">&#40;</span><span style="color: #008000;">Exception</span><span style="color: black;">&#41;</span>
    <span style="color: #ff7700;font-weight:bold;">pass</span>
&nbsp;
<span style="color: #ff7700;font-weight:bold;">class</span> MonTraitementError<span style="color: black;">&#40;</span>MonProjetError<span style="color: black;">&#41;</span>
    <span style="color: #ff7700;font-weight:bold;">pass</span></pre></td></tr></table></div>

<p>C&#8217;est généralement une bonne idée quand vous créez une lib et que vous voulez permettre aux utilisateurs de votre bibliothèque de pouvoir filtrer les exceptions levées par votre code uniquement.</p>
<p>Néanmoins, dans ce cas, choisissez intelligemment les parents de vos exceptions personnalisées. Chaque exception de Python est porteuse de sens. <code>TypeError</code> est levée pour une erreur de type, <code>IOError</code> pour une erreur d&#8217;entrée/sortie, etc. Il est logique qu&#8217;un développeur s&#8217;attende à pouvoir les attraper selon ce sens.</p>
<p>Heureusement, on peut hériter de plusieurs exceptions :</p>

<div class="wp_syntax"><table><tr><td class="code"><pre class="python" style="font-family:monospace;"><span style="color: #ff7700;font-weight:bold;">class</span> MonProjetError<span style="color: black;">&#40;</span><span style="color: #008000;">Exception</span><span style="color: black;">&#41;</span>
    <span style="color: #ff7700;font-weight:bold;">pass</span>
&nbsp;
<span style="color: #ff7700;font-weight:bold;">class</span> MonTraitementError<span style="color: black;">&#40;</span>MonProjetError<span style="color: #66cc66;">,</span> <span style="color: #008000;">TypeError</span><span style="color: black;">&#41;</span>
    <span style="color: #ff7700;font-weight:bold;">pass</span></pre></td></tr></table></div>

<p>Du coup, si quelqu&#8217;un essaye d&#8217;attraper <code>MonTraitementError</code> ou <code>TypeError</code> ou <code>MonProjetError</code> selon ce qu&#8217;il désire exprimer, ça marchera.</p>
<p>En effet, les exceptions sont une forme d&#8217;expression :</p>
<ul>
<li>Celui qui lève l&#8217;exception dit explicitement ce qui peut merder : quand on lit son code, on comprend les cas d&#8217;erreurs.</li>
<li>Celui qui l&#8217;attrape dit explicitement ce qu&#8217;il veut gérer : quand on lit son code, on comprend l&#8217;objectif.</li>
<li>Python dit explicitement ce qui a foiré. Quand on lit la stack trace, on comprend ce qui a merdé.</li>
</ul>
<p><strong>Vous communiquez en utilisant des exceptions entre développeurs et utilisateurs du code.</strong></p>
<h2>Finally et else</h2>
<p><code>try/except</code> peut être complété par deux autres mots clés : <code>finally</code> et <code>else</code>.</p>
<p><code>else</code> est le bloc exécuté si aucune exception n&#8217;est levée :</p>

<div class="wp_syntax"><table><tr><td class="code"><pre class="python" style="font-family:monospace;">    <span style="color: #ff7700;font-weight:bold;">try</span>:
        <span style="color: #ff7700;font-weight:bold;">return</span> personnages<span style="color: black;">&#91;</span><span style="color: #ff4500;">100</span> / i<span style="color: black;">&#93;</span>
    <span style="color: #ff7700;font-weight:bold;">except</span> <span style="color: black;">&#40;</span><span style="color: #008000;">IndexError</span><span style="color: #66cc66;">,</span> <span style="color: #008000;">ZeroDivisionError</span><span style="color: black;">&#41;</span>:
        <span style="color: #ff7700;font-weight:bold;">return</span> <span style="color: #008000;">None</span>
    <span style="color: #ff7700;font-weight:bold;">except</span> <span style="color: #008000;">KeyError</span>:
        <span style="color: #ff7700;font-weight:bold;">print</span><span style="color: black;">&#40;</span><span style="color: #483d8b;">&quot;Qui est le bâtard qui a remplacé ma &quot;</span>
              <span style="color: #483d8b;">&quot;liste par un dico dans mon dos ?&quot;</span><span style="color: black;">&#41;</span>
    <span style="color: #ff7700;font-weight:bold;">else</span>:
        <span style="color: #ff7700;font-weight:bold;">print</span><span style="color: black;">&#40;</span><span style="color: #483d8b;">'Bon en fait tout va bien'</span><span style="color: black;">&#41;</span></pre></td></tr></table></div>

<p><code>finally</code> est un bloc qui est exécuté après que tous les autres blocs aient été exécutés, peu importe qu&#8217;il y ait eu une exception ou non, et <strong>même si le programme crash</strong>.</p>

<div class="wp_syntax"><table><tr><td class="code"><pre class="python" style="font-family:monospace;">    <span style="color: #ff7700;font-weight:bold;">try</span>:
        <span style="color: #ff7700;font-weight:bold;">return</span> personnages<span style="color: black;">&#91;</span><span style="color: #ff4500;">100</span> / i<span style="color: black;">&#93;</span>
    <span style="color: #ff7700;font-weight:bold;">except</span> <span style="color: black;">&#40;</span><span style="color: #008000;">IndexError</span><span style="color: #66cc66;">,</span> <span style="color: #008000;">ZeroDivisionError</span><span style="color: black;">&#41;</span>:
        <span style="color: #ff7700;font-weight:bold;">return</span> <span style="color: #008000;">None</span>
    <span style="color: #ff7700;font-weight:bold;">except</span> <span style="color: #008000;">KeyError</span>:
        <span style="color: #ff7700;font-weight:bold;">print</span><span style="color: black;">&#40;</span><span style="color: #483d8b;">&quot;Qui est le bâtard qui a remplacé ma &quot;</span>
              <span style="color: #483d8b;">&quot;liste par un dico dans mon dos ?&quot;</span><span style="color: black;">&#41;</span>
    <span style="color: #ff7700;font-weight:bold;">finally</span>:
        <span style="color: #ff7700;font-weight:bold;">print</span><span style="color: black;">&#40;</span><span style="color: #483d8b;">'Rosebud !'</span><span style="color: black;">&#41;</span></pre></td></tr></table></div>

<p>Dans notre cas, peu importe ce qui se passe dans le bloc <code>try/except</code>, <code>Rosebud !</code> est toujours affiché.</p>
<p><code>finally</code> n&#8217;est pas à l&#8217;épreuve des balles, si on crash l’interpréteur violemment (ou que vous débranchez la prise du serveur :)), il ne vous sauvera pas. Mais il permet de gérer la plupart des cas extrêmes.</p>
<p>Ah, oui, au fait, si vous utilisez <code>finally</code>, <code>except</code> n&#8217;est pas obligatoire.</p>
<h2>Quand utiliser try, except, else et finally ?</h2>
<p>On utilisera <code>try/except</code> d&#8217;abord et avant tout pour gérer les erreurs sur laquelle on a pas de controle.</p>
<p>Par exemple, vous téléchargez une document depuis le Web :</p>

<div class="wp_syntax"><table><tr><td class="code"><pre class="python" style="font-family:monospace;"><span style="color: #66cc66;">&gt;&gt;&gt;</span> <span style="color: #ff7700;font-weight:bold;">import</span> <span style="color: #dc143c;">urllib</span>.<span style="color: black;">request</span>
<span style="color: #66cc66;">&gt;&gt;&gt;</span> fichier_recu<span style="color: #66cc66;">,</span> headers <span style="color: #66cc66;">=</span> <span style="color: #dc143c;">urllib</span>.<span style="color: black;">request</span>.<span style="color: black;">urlretrieve</span><span style="color: black;">&#40;</span><span style="color: #483d8b;">'http://www.sudinfo.be/sites/default/files/imagecache/pagallery_450x300/1365338424_B9765144Z.1_20121219124926_000_GOJDROM6.1-0.png'</span><span style="color: black;">&#41;</span></pre></td></tr></table></div>

<p>Ca va marcher la plupart du temps. Et puis un jour, le réseau va sauter, et ça va plus marcher :</p>

<div class="wp_syntax"><table><tr><td class="code"><pre class="python" style="font-family:monospace;"><span style="color: #66cc66;">&gt;&gt;&gt;</span> fichier_recu<span style="color: #66cc66;">,</span> headers <span style="color: #66cc66;">=</span> <span style="color: #dc143c;">urllib</span>.<span style="color: black;">request</span>.<span style="color: black;">urlretrieve</span><span style="color: black;">&#40;</span><span style="color: #483d8b;">'http://www.sudinfo.be/sites/default/files/imagecache/pagallery_450x300/1365338424_B9765144Z.1_20121219124926_000_GOJDROM6.1-0.png'</span><span style="color: black;">&#41;</span>
Traceback <span style="color: black;">&#40;</span>most recent call last<span style="color: black;">&#41;</span>:
  File <span style="color: #483d8b;">&quot;&lt;ipython-input-10-0f41d15585f0&gt;&quot;</span><span style="color: #66cc66;">,</span> line <span style="color: #ff4500;">1</span><span style="color: #66cc66;">,</span> <span style="color: #ff7700;font-weight:bold;">in</span> <span style="color: #66cc66;">&lt;</span>module<span style="color: #66cc66;">&gt;</span>
    fichier_recu<span style="color: #66cc66;">,</span> headers <span style="color: #66cc66;">=</span> <span style="color: #dc143c;">urllib</span>.<span style="color: black;">request</span>.<span style="color: black;">urlretrieve</span><span style="color: black;">&#40;</span><span style="color: #483d8b;">'http://www.sudinfo.be/sites/default/files/imagecache/pagallery_450x300/1365338424_B9765144Z.1_20121219124926_000_GOJDROM6.1-0.png'</span><span style="color: black;">&#41;</span>
  File <span style="color: #483d8b;">&quot;/usr/lib/python3.4/urllib/request.py&quot;</span><span style="color: #66cc66;">,</span> line <span style="color: #ff4500;">178</span><span style="color: #66cc66;">,</span> <span style="color: #ff7700;font-weight:bold;">in</span> urlretrieve
    <span style="color: #ff7700;font-weight:bold;">with</span> contextlib.<span style="color: black;">closing</span><span style="color: black;">&#40;</span>urlopen<span style="color: black;">&#40;</span>url<span style="color: #66cc66;">,</span> data<span style="color: black;">&#41;</span><span style="color: black;">&#41;</span> <span style="color: #ff7700;font-weight:bold;">as</span> fp:
  File <span style="color: #483d8b;">&quot;/usr/lib/python3.4/urllib/request.py&quot;</span><span style="color: #66cc66;">,</span> line <span style="color: #ff4500;">153</span><span style="color: #66cc66;">,</span> <span style="color: #ff7700;font-weight:bold;">in</span> urlopen
    <span style="color: #ff7700;font-weight:bold;">return</span> opener.<span style="color: #008000;">open</span><span style="color: black;">&#40;</span>url<span style="color: #66cc66;">,</span> data<span style="color: #66cc66;">,</span> timeout<span style="color: black;">&#41;</span>
  File <span style="color: #483d8b;">&quot;/usr/lib/python3.4/urllib/request.py&quot;</span><span style="color: #66cc66;">,</span> line <span style="color: #ff4500;">455</span><span style="color: #66cc66;">,</span> <span style="color: #ff7700;font-weight:bold;">in</span> <span style="color: #008000;">open</span>
    response <span style="color: #66cc66;">=</span> <span style="color: #008000;">self</span>._open<span style="color: black;">&#40;</span>req<span style="color: #66cc66;">,</span> data<span style="color: black;">&#41;</span>
  File <span style="color: #483d8b;">&quot;/usr/lib/python3.4/urllib/request.py&quot;</span><span style="color: #66cc66;">,</span> line <span style="color: #ff4500;">473</span><span style="color: #66cc66;">,</span> <span style="color: #ff7700;font-weight:bold;">in</span> _open
    <span style="color: #483d8b;">'_open'</span><span style="color: #66cc66;">,</span> req<span style="color: black;">&#41;</span>
  File <span style="color: #483d8b;">&quot;/usr/lib/python3.4/urllib/request.py&quot;</span><span style="color: #66cc66;">,</span> line <span style="color: #ff4500;">433</span><span style="color: #66cc66;">,</span> <span style="color: #ff7700;font-weight:bold;">in</span> _call_chain
    result <span style="color: #66cc66;">=</span> func<span style="color: black;">&#40;</span>*args<span style="color: black;">&#41;</span>
  File <span style="color: #483d8b;">&quot;/usr/lib/python3.4/urllib/request.py&quot;</span><span style="color: #66cc66;">,</span> line <span style="color: #ff4500;">1258</span><span style="color: #66cc66;">,</span> <span style="color: #ff7700;font-weight:bold;">in</span> http_open
    <span style="color: #ff7700;font-weight:bold;">return</span> <span style="color: #008000;">self</span>.<span style="color: black;">do_open</span><span style="color: black;">&#40;</span>http.<span style="color: black;">client</span>.<span style="color: black;">HTTPConnection</span><span style="color: #66cc66;">,</span> req<span style="color: black;">&#41;</span>
  File <span style="color: #483d8b;">&quot;/usr/lib/python3.4/urllib/request.py&quot;</span><span style="color: #66cc66;">,</span> line <span style="color: #ff4500;">1235</span><span style="color: #66cc66;">,</span> <span style="color: #ff7700;font-weight:bold;">in</span> do_open
    <span style="color: #ff7700;font-weight:bold;">raise</span> URLError<span style="color: black;">&#40;</span>err<span style="color: black;">&#41;</span>
URLError: <span style="color: #66cc66;">&lt;</span>urlopen error <span style="color: black;">&#91;</span>Errno -<span style="color: #ff4500;">2</span><span style="color: black;">&#93;</span> Name <span style="color: #ff7700;font-weight:bold;">or</span> service <span style="color: #ff7700;font-weight:bold;">not</span> known<span style="color: #66cc66;">&gt;</span></pre></td></tr></table></div>

<p>On va donc gérer l&#8217;exception (et beaucoup d&#8217;autres, le réseau ça foire tout le temps) :</p>

<div class="wp_syntax"><table><tr><td class="code"><pre class="python" style="font-family:monospace;">&nbsp;
<span style="color: #ff7700;font-weight:bold;">try</span>:
    fichier_recu<span style="color: #66cc66;">,</span> headers <span style="color: #66cc66;">=</span> <span style="color: #dc143c;">urllib</span>.<span style="color: black;">request</span>.<span style="color: black;">urlretrieve</span><span style="color: black;">&#40;</span><span style="color: #483d8b;">'http://www.sudinfo.be/sites/default/files/imagecache/pagallery_450x300/1365338424_B9765144Z.1_20121219124926_000_GOJDROM6.1-0.png'</span><span style="color: black;">&#41;</span>
<span style="color: #ff7700;font-weight:bold;">except</span> <span style="color: #dc143c;">urllib</span>.<span style="color: black;">URLError</span>:
    <span style="color: #808080; font-style: italic;"># faire un truc</span></pre></td></tr></table></div>

<p>Parmi les choses à faire : réessayer plus tard, tenter de redémarrer la carte réseau, changer d&#8217;URL&#8230;</p>
<p>Un autre raison d&#8217;utiliser <code>try/except</code> est pour enregistrer une trace des erreurs.</p>

<div class="wp_syntax"><table><tr><td class="code"><pre class="python" style="font-family:monospace;">&nbsp;
<span style="color: #ff7700;font-weight:bold;">try</span>:
    url <span style="color: #66cc66;">=</span> <span style="color: #483d8b;">'http://www.sudinfo.be/sites/default/files/imagecache/pagallery_450x300/1365338424_B9765144Z.1_20121219124926_000_GOJDROM6.1-0.png'</span>
    fichier_recu<span style="color: #66cc66;">,</span> headers <span style="color: #66cc66;">=</span> <span style="color: #dc143c;">urllib</span>.<span style="color: black;">request</span>.<span style="color: black;">urlretrieve</span><span style="color: black;">&#40;</span>url<span style="color: black;">&#41;</span>
<span style="color: #ff7700;font-weight:bold;">except</span> <span style="color: #dc143c;">urllib</span>.<span style="color: black;">URLError</span>:
    <span style="color: #ff7700;font-weight:bold;">with</span> <span style="color: #008000;">open</span><span style="color: black;">&#40;</span><span style="color: #483d8b;">'/tmp/errors.log'</span><span style="color: #66cc66;">,</span> <span style="color: #483d8b;">'a'</span><span style="color: black;">&#41;</span> <span style="color: #ff7700;font-weight:bold;">as</span> f:
        f.<span style="color: black;">write</span><span style="color: black;">&#40;</span><span style="color: #483d8b;">&quot;L'URL '%s' est inacessible&quot;</span> % url<span style="color: black;">&#41;</span></pre></td></tr></table></div>

<p>On peut le faire avec le module <code>logging</code>, ou envoyer un mail ou un SMS d&#8217;alerte&#8230;</p>
<p>Proposer un comportement par défaut est aussi très courant.</p>
<p>Par exemple, vous voulez récupérer le premier élément d&#8217;une liste, mais si la liste est vide, récupérer <code>None</code> :</p>

<div class="wp_syntax"><table><tr><td class="code"><pre class="python" style="font-family:monospace;"><span style="color: #ff7700;font-weight:bold;">try</span>:
    resultat <span style="color: #66cc66;">=</span> liste<span style="color: black;">&#91;</span><span style="color: #ff4500;">0</span><span style="color: black;">&#93;</span>
<span style="color: #ff7700;font-weight:bold;">except</span> <span style="color: #008000;">IndexError</span>:
    resultat <span style="color: #66cc66;">=</span> <span style="color: #008000;">None</span></pre></td></tr></table></div>

<p>Ce cas d&#8217;usage est typique de Python. Dans d&#8217;autres langage, on regarderait si le tableau est vide :</p>

<div class="wp_syntax"><table><tr><td class="code"><pre class="python" style="font-family:monospace;"><span style="color: #ff7700;font-weight:bold;">if</span> liste:
    resultat <span style="color: #66cc66;">=</span> liste<span style="color: black;">&#91;</span><span style="color: #ff4500;">0</span><span style="color: black;">&#93;</span>
<span style="color: #ff7700;font-weight:bold;">else</span>:
    resultat <span style="color: #66cc66;">=</span> <span style="color: #008000;">None</span></pre></td></tr></table></div>

<p>C&#8217;est typiquement la philosophie Java : &#8220;Watch before you leap&#8221; (WBYL). Soit &#8220;Regardez où vous mettez les pieds&#8221;.</p>
<p>En Python, la philosophie est &#8220;It&#8217;s easier to ask for forgiveness than permission&#8221; (EAFP). Soit &#8220;il est plus facile de demander pardon que la permission&#8221;.</p>
<p>La raison est que les exceptions sont très rapides en Python, et permettent de gérer plein de cas d&#8217;un coup.</p>
<p>Par exemple, dans le cas :</p>

<div class="wp_syntax"><table><tr><td class="code"><pre class="python" style="font-family:monospace;"><span style="color: #ff7700;font-weight:bold;">if</span> conteneur:
    resultat <span style="color: #66cc66;">=</span> conteneur<span style="color: black;">&#91;</span><span style="color: #ff4500;">0</span><span style="color: black;">&#93;</span>
<span style="color: #ff7700;font-weight:bold;">else</span>:
    resultat <span style="color: #66cc66;">=</span> <span style="color: #008000;">None</span></pre></td></tr></table></div>

<p>Que se passe-t-il si conteneur peut être un dictionnaire ou une liste ? C&#8217;est possible en Python du fait du <a href="http://sametmax.com/quest-ce-que-le-duck-typing-et-a-quoi-ca-sert/">duck typing</a>.</p>
<p>Il faut tester deux cas :</p>

<div class="wp_syntax"><table><tr><td class="code"><pre class="python" style="font-family:monospace;"><span style="color: #ff7700;font-weight:bold;">if</span> conteneur <span style="color: #ff7700;font-weight:bold;">and</span> <span style="color: #ff4500;">0</span> <span style="color: #ff7700;font-weight:bold;">in</span> conteneur:
    resultat <span style="color: #66cc66;">=</span> conteneur<span style="color: black;">&#91;</span><span style="color: #ff4500;">0</span><span style="color: black;">&#93;</span>
<span style="color: #ff7700;font-weight:bold;">else</span>:
    resultat <span style="color: #66cc66;">=</span> <span style="color: #008000;">None</span></pre></td></tr></table></div>

<p>Avec une exception, on teste un seul cas :</p>

<div class="wp_syntax"><table><tr><td class="code"><pre class="python" style="font-family:monospace;"><span style="color: #ff7700;font-weight:bold;">try</span>:
    resultat <span style="color: #66cc66;">=</span> conteneur<span style="color: black;">&#91;</span><span style="color: #ff4500;">0</span><span style="color: black;">&#93;</span>
<span style="color: #ff7700;font-weight:bold;">except</span> <span style="color: #008000;">LookupError</span>:
    resultat <span style="color: #66cc66;">=</span> <span style="color: #008000;">None</span></pre></td></tr></table></div>

<p>Là c&#8217;est un cas simple, mais sur des cas complexes comme les fichiers, c&#8217;est très utile :</p>

<div class="wp_syntax"><table><tr><td class="code"><pre class="python" style="font-family:monospace;"><span style="color: #ff7700;font-weight:bold;">try</span>:
    fichier <span style="color: #66cc66;">=</span> <span style="color: #008000;">open</span><span style="color: black;">&#40;</span><span style="color: #483d8b;">'/tmp/fichier'</span><span style="color: #66cc66;">,</span> <span style="color: #483d8b;">'w'</span><span style="color: black;">&#41;</span>
<span style="color: #ff7700;font-weight:bold;">except</span> <span style="color: black;">&#40;</span><span style="color: #008000;">IOError</span><span style="color: #66cc66;">,</span> <span style="color: #008000;">OSError</span><span style="color: black;">&#41;</span>:
    <span style="color: #808080; font-style: italic;"># gérer l'erreur</span></pre></td></tr></table></div>

<p>Si je devais faire ça avec des <code>if</code>, il faudrait :</p>
<ul>
<li>Vérifier que le fichier existe.</li>
<li>Vérifier que le fichier est un fichier et non un dossier.</li>
<li>Vérifier que j&#8217;ai les permissions d&#8217;écrire sur le fichier.</li>
<li>Vérifier que j&#8217;ai les permissions de traverser les dossiers parents.</li>
<li>Vérifier que personne n&#8217;a ouvert le fichier en écriture avant.</li>
</ul>
<p>C&#8217;est relou. Mais en prime, entre chaque vérification du <code>if</code> s&#8217;écoule quelques nanosecondes durant lequelles l&#8217;état de mon fichier a pu changer. Un autre programme a pu changer les permissions, créer un fichier avec le même nom, etc. Arrivé à la fin du <code>if</code>, on n&#8217;a pas la garantie que les vérifications du début du <code>if</code> sont toujours d&#8217;actualité.</p>
<p><code>try/except</code> évite ce problème qu&#8217;on appelle des &#8220;race conditions&#8221; : on essaye d&#8217;abord, et ensuite si ça foire, on analyse pourquoi. Comme en drague.</p>
<p>Enfin on va utiliser <code>finally</code> pour faire du nettoyage : fermer un fichier, fermer une connexion à une base de données, une socket, supprimer un fichier temporaire. Bref, tout ce qu&#8217;on veut qui arrive, même si notre code se vautre.</p>

<div class="wp_syntax"><table><tr><td class="code"><pre class="python" style="font-family:monospace;"><span style="color: #ff7700;font-weight:bold;">try</span>:
    fichier <span style="color: #66cc66;">=</span> <span style="color: #008000;">open</span><span style="color: black;">&#40;</span><span style="color: #483d8b;">'/tmp/fichier'</span><span style="color: #66cc66;">,</span> <span style="color: #483d8b;">'w'</span><span style="color: black;">&#41;</span>
<span style="color: #ff7700;font-weight:bold;">except</span> <span style="color: black;">&#40;</span><span style="color: #008000;">IOError</span><span style="color: #66cc66;">,</span> <span style="color: #008000;">OSError</span><span style="color: black;">&#41;</span>:
    <span style="color: #808080; font-style: italic;"># gérer l'erreur</span>
<span style="color: #ff7700;font-weight:bold;">else</span>:
    <span style="color: #808080; font-style: italic;"># faire un truc avec le fichier</span>
&nbsp;
<span style="color: #808080; font-style: italic;"># on essaye toujours de fermer notre fichier</span>
<span style="color: #ff7700;font-weight:bold;">finally</span>:
    <span style="color: #ff7700;font-weight:bold;">try</span>:
        fichier.<span style="color: black;">close</span><span style="color: black;">&#40;</span><span style="color: black;">&#41;</span>
&nbsp;
    <span style="color: #808080; font-style: italic;"># le fichier n'a jamais été ouvert et</span>
    <span style="color: #808080; font-style: italic;"># la variable n'existe pas</span>
    <span style="color: #ff7700;font-weight:bold;">except</span> <span style="color: #008000;">NameError</span>:
        <span style="color: #ff7700;font-weight:bold;">pass</span></pre></td></tr></table></div>

<h2>with</h2>
<p>Comme vous avez pu le constater, gérer proprement les erreurs, ça peut devenir vite chiant. Pour cette raison, beaucoup de <a href="http://sametmax.com/les-context-managers-et-le-mot-cle-with-en-python/">context managers</a> servent à gérer les cas les plus courants.</p>
<p>Typiquement :</p>

<div class="wp_syntax"><table><tr><td class="code"><pre class="python" style="font-family:monospace;"><span style="color: #ff7700;font-weight:bold;">try</span>:
    fichier <span style="color: #66cc66;">=</span> <span style="color: #008000;">open</span><span style="color: black;">&#40;</span><span style="color: #483d8b;">'/tmp/fichier'</span><span style="color: #66cc66;">,</span> <span style="color: #483d8b;">'w'</span><span style="color: black;">&#41;</span>
<span style="color: #ff7700;font-weight:bold;">except</span> <span style="color: black;">&#40;</span><span style="color: #008000;">IOError</span><span style="color: #66cc66;">,</span> <span style="color: #008000;">OSError</span><span style="color: black;">&#41;</span>:
    <span style="color: #808080; font-style: italic;"># gérer l'erreur</span>
<span style="color: #ff7700;font-weight:bold;">else</span>:
    <span style="color: #808080; font-style: italic;"># faire un truc avec le fichier</span>
<span style="color: #ff7700;font-weight:bold;">finally</span>:
    <span style="color: #ff7700;font-weight:bold;">try</span>:
        fichier.<span style="color: black;">close</span><span style="color: black;">&#40;</span><span style="color: black;">&#41;</span>
    <span style="color: #ff7700;font-weight:bold;">except</span> <span style="color: #008000;">NameError</span>:
        <span style="color: #ff7700;font-weight:bold;">pass</span></pre></td></tr></table></div>

<p>Peut être remplacé par :</p>

<div class="wp_syntax"><table><tr><td class="code"><pre class="python" style="font-family:monospace;"><span style="color: #ff7700;font-weight:bold;">try</span>:
    <span style="color: #ff7700;font-weight:bold;">with</span> <span style="color: #008000;">open</span><span style="color: black;">&#40;</span><span style="color: #483d8b;">'/tmp/fichier'</span><span style="color: #66cc66;">,</span> <span style="color: #483d8b;">'w'</span><span style="color: black;">&#41;</span> <span style="color: #ff7700;font-weight:bold;">as</span> fichier:
        <span style="color: #808080; font-style: italic;"># faire un truc avec le fichier</span>
<span style="color: #ff7700;font-weight:bold;">except</span> <span style="color: black;">&#40;</span><span style="color: #008000;">IOError</span><span style="color: #66cc66;">,</span> <span style="color: #008000;">OSError</span><span style="color: black;">&#41;</span>:
    <span style="color: #808080; font-style: italic;"># gérer l'erreur</span></pre></td></tr></table></div>

<p>Ou même en Python 3:</p>

<div class="wp_syntax"><table><tr><td class="code"><pre class="python" style="font-family:monospace;"><span style="color: #ff7700;font-weight:bold;">try</span>:
    <span style="color: #ff7700;font-weight:bold;">with</span> <span style="color: #008000;">open</span><span style="color: black;">&#40;</span><span style="color: #483d8b;">'/tmp/fichier'</span><span style="color: #66cc66;">,</span> <span style="color: #483d8b;">'w'</span><span style="color: black;">&#41;</span> <span style="color: #ff7700;font-weight:bold;">as</span> fichier:
        <span style="color: #808080; font-style: italic;"># faire un truc avec le fichier</span>
<span style="color: #ff7700;font-weight:bold;">except</span> <span style="color: #008000;">EnvironmentError</span>:
    <span style="color: #808080; font-style: italic;"># gérer l'erreur</span></pre></td></tr></table></div>

<p>Ce qui est quand même vachement plus court. Le context manager va s&#8217;occuper de la fermeture du fichier pour nous proprement. Cet article est déjà long, donc je ne vais pas rentrer en détail sur <code>with</code>, mais sachez que beaucoup d&#8217;objets qui demandent une fermeture peuvent être traités comme ça.</p>
<h2>Bubbling</h2>
<p>Une chose qui échappe généralement aux débutants quand ils commencent à utiliser les exceptions, c&#8217;est qu&#8217;elles bubblent, c&#8217;est à dire qu&#8217;elles remontent comme une bulle, de bloc en bloc, jusqu&#8217;à la racine du programme.</p>
<p>Si j&#8217;ai :</p>

<div class="wp_syntax"><table><tr><td class="code"><pre class="python" style="font-family:monospace;"><span style="color: #ff7700;font-weight:bold;">if</span> truc:
    <span style="color: #ff7700;font-weight:bold;">if</span> machin:
        <span style="color: #ff7700;font-weight:bold;">for</span> bidule <span style="color: #ff7700;font-weight:bold;">in</span> chose:
            <span style="color: #ff7700;font-weight:bold;">raise</span> MaisAieuuuuuError<span style="color: black;">&#40;</span><span style="color: #483d8b;">&quot;Ca fait mal t'es con&quot;</span><span style="color: black;">&#41;</span></pre></td></tr></table></div>

<p>Mon programme ne va PAS planter à la ligne <code>raise MaisAieuuuuuError("Ca fait mal t'es con")</code>.</p>
<p>A la place, l&#8217;erreur va remonter en dehors de la boucle. Si elle n&#8217;est pas attrapée, elle va remonter en dehors du premier <code>if</code>. Si elle n&#8217;est pas attrapée, elle va remonter en dehors du second <code>if</code>, et enfin, si elle vraiment, mais alors vraiment pas attrapée, Python va crasher, et montrer une belle stack trace.</p>
<p>Ça veut dire que vous pouvez arrêter l&#8217;exception à plusieurs endroits dans le programme :</p>

<div class="wp_syntax"><table><tr><td class="code"><pre class="python" style="font-family:monospace;"><span style="color: #ff7700;font-weight:bold;">if</span> truc:
    <span style="color: #ff7700;font-weight:bold;">if</span> machin:
        <span style="color: #ff7700;font-weight:bold;">for</span> bidule <span style="color: #ff7700;font-weight:bold;">in</span> chose:
            <span style="color: #ff7700;font-weight:bold;">try</span>:
                <span style="color: #ff7700;font-weight:bold;">raise</span> MaisAieuuuuuError<span style="color: black;">&#40;</span><span style="color: #483d8b;">&quot;Ca fait mal t'es con&quot;</span><span style="color: black;">&#41;</span>
            <span style="color: #ff7700;font-weight:bold;">except</span> MaisAieuuuuuError:
                <span style="color: #808080; font-style: italic;"># faire un truc</span></pre></td></tr></table></div>

<p>Dans ce cas, la boucle va continuer à tourner après la première erreur, et faire un <code>try/except</code> à chaque tour de boucle.</p>
<p>Mais si vous faites :</p>

<div class="wp_syntax"><table><tr><td class="code"><pre class="python" style="font-family:monospace;"><span style="color: #ff7700;font-weight:bold;">if</span> truc:
    <span style="color: #ff7700;font-weight:bold;">if</span> machin:
        <span style="color: #ff7700;font-weight:bold;">try</span>
            <span style="color: #ff7700;font-weight:bold;">for</span> bidule <span style="color: #ff7700;font-weight:bold;">in</span> chose:
                <span style="color: #ff7700;font-weight:bold;">raise</span> MaisAieuuuuuError<span style="color: black;">&#40;</span><span style="color: #483d8b;">&quot;Ca fait mal t'es con&quot;</span><span style="color: black;">&#41;</span>
        <span style="color: #ff7700;font-weight:bold;">except</span> MaisAieuuuuuError:
            <span style="color: #808080; font-style: italic;"># faire un truc</span></pre></td></tr></table></div>

<p>A la première erreur, la boucle est interrompue définitivement. Par contre la condition <code>if machin</code> va continuer sa route.</p>
<p>Mais si vous faites :</p>

<div class="wp_syntax"><table><tr><td class="code"><pre class="python" style="font-family:monospace;"><span style="color: #ff7700;font-weight:bold;">if</span> truc:
    <span style="color: #ff7700;font-weight:bold;">try</span>:
        <span style="color: #ff7700;font-weight:bold;">if</span> machin:
            <span style="color: #ff7700;font-weight:bold;">for</span> bidule <span style="color: #ff7700;font-weight:bold;">in</span> chose:
                <span style="color: #ff7700;font-weight:bold;">raise</span> MaisAieuuuuuError<span style="color: black;">&#40;</span><span style="color: #483d8b;">&quot;Ca fait mal t'es con&quot;</span><span style="color: black;">&#41;</span>
    <span style="color: #ff7700;font-weight:bold;">except</span> MaisAieuuuuuError:
        <span style="color: #808080; font-style: italic;"># faire un truc</span></pre></td></tr></table></div>

<p>La condition <code>if machin</code> est définitivement interrompue.</p>
<p>Et si vous faites :</p>

<div class="wp_syntax"><table><tr><td class="code"><pre class="python" style="font-family:monospace;"><span style="color: #ff7700;font-weight:bold;">try</span>:
    <span style="color: #ff7700;font-weight:bold;">if</span> truc:
        <span style="color: #ff7700;font-weight:bold;">if</span> machin:
            <span style="color: #ff7700;font-weight:bold;">for</span> bidule <span style="color: #ff7700;font-weight:bold;">in</span> chose:
                <span style="color: #ff7700;font-weight:bold;">raise</span> MaisAieuuuuuError<span style="color: black;">&#40;</span><span style="color: #483d8b;">&quot;Ca fait mal t'es con&quot;</span><span style="color: black;">&#41;</span>
<span style="color: #ff7700;font-weight:bold;">except</span> MaisAieuuuuuError:
    <span style="color: #808080; font-style: italic;"># faire un truc</span></pre></td></tr></table></div>

<p>Là, tout le programme est interrompu. Ca ne plantera pas, mais ça ne fera plus rien.</p>
<p>Il faut donc choisir soigneusement où attraper l&#8217;exception selon la partie du programme qu&#8217;on souhaite interrompre.</p>
<h2>Manipuler l&#8217;exception attrapée</h2>
<p>Les exceptions sont des objets complets. On peut les récupérer avec le mot clés <code>as</code> et analyser leur contenu.</p>
<p>Exemple :</p>

<div class="wp_syntax"><table><tr><td class="code"><pre class="python" style="font-family:monospace;"><span style="color: #ff7700;font-weight:bold;">try</span>:
    f <span style="color: #66cc66;">=</span> <span style="color: #008000;">open</span><span style="color: black;">&#40;</span><span style="color: #483d8b;">&quot;nawak&quot;</span><span style="color: black;">&#41;</span>
<span style="color: #ff7700;font-weight:bold;">except</span> <span style="color: #008000;">IOError</span> <span style="color: #ff7700;font-weight:bold;">as</span> e:
    <span style="color: #ff7700;font-weight:bold;">print</span><span style="color: black;">&#40;</span><span style="color: #483d8b;">&quot;args: &quot;</span><span style="color: #66cc66;">,</span> e.<span style="color: black;">args</span><span style="color: black;">&#41;</span>
    <span style="color: #ff7700;font-weight:bold;">print</span><span style="color: black;">&#40;</span><span style="color: #483d8b;">&quot;errno: &quot;</span><span style="color: #66cc66;">,</span> e.<span style="color: #dc143c;">errno</span><span style="color: black;">&#41;</span>
    <span style="color: #ff7700;font-weight:bold;">print</span><span style="color: black;">&#40;</span><span style="color: #483d8b;">&quot;filename: &quot;</span><span style="color: #66cc66;">,</span> e.<span style="color: black;">filename</span><span style="color: black;">&#41;</span>
    <span style="color: #ff7700;font-weight:bold;">print</span><span style="color: black;">&#40;</span><span style="color: #483d8b;">&quot;strerror: &quot;</span><span style="color: #66cc66;">,</span> e.<span style="color: black;">strerror</span><span style="color: black;">&#41;</span>
&nbsp;
<span style="color: #808080; font-style: italic;">## args:  (2, 'No such file or directory')</span>
<span style="color: #808080; font-style: italic;">## errno:  2</span>
<span style="color: #808080; font-style: italic;">## filename:  nawak</span>
<span style="color: #808080; font-style: italic;">## strerror:  No such file or directory</span></pre></td></tr></table></div>

<p>Il est même possible de les lever à nouveau après :</p>

<div class="wp_syntax"><table><tr><td class="code"><pre class="python" style="font-family:monospace;"><span style="color: #ff7700;font-weight:bold;">try</span>:
    f <span style="color: #66cc66;">=</span> <span style="color: #008000;">open</span><span style="color: black;">&#40;</span><span style="color: #483d8b;">&quot;nawak&quot;</span><span style="color: black;">&#41;</span>
<span style="color: #ff7700;font-weight:bold;">except</span> <span style="color: #008000;">IOError</span> <span style="color: #ff7700;font-weight:bold;">as</span> e:
    <span style="color: #ff7700;font-weight:bold;">print</span><span style="color: black;">&#40;</span><span style="color: #483d8b;">&quot;args: &quot;</span><span style="color: #66cc66;">,</span> e.<span style="color: black;">args</span><span style="color: black;">&#41;</span>
    <span style="color: #ff7700;font-weight:bold;">print</span><span style="color: black;">&#40;</span><span style="color: #483d8b;">&quot;errno: &quot;</span><span style="color: #66cc66;">,</span> e.<span style="color: #dc143c;">errno</span><span style="color: black;">&#41;</span>
    <span style="color: #ff7700;font-weight:bold;">print</span><span style="color: black;">&#40;</span><span style="color: #483d8b;">&quot;filename: &quot;</span><span style="color: #66cc66;">,</span> e.<span style="color: black;">filename</span><span style="color: black;">&#41;</span>
    <span style="color: #ff7700;font-weight:bold;">print</span><span style="color: black;">&#40;</span><span style="color: #483d8b;">&quot;strerror: &quot;</span><span style="color: #66cc66;">,</span> e.<span style="color: black;">strerror</span><span style="color: black;">&#41;</span>
    <span style="color: #ff7700;font-weight:bold;">raise</span> e</pre></td></tr></table></div>

<p><code>raise</code> appelé sans argument lève de toute façon automatiquement l&#8217;exception en cours. Mais il est possible de lever une autre exception :</p>

<div class="wp_syntax"><table><tr><td class="code"><pre class="python" style="font-family:monospace;"><span style="color: #ff7700;font-weight:bold;">try</span>:
    f <span style="color: #66cc66;">=</span> <span style="color: #008000;">open</span><span style="color: black;">&#40;</span><span style="color: #483d8b;">&quot;nawak&quot;</span><span style="color: black;">&#41;</span>
<span style="color: #ff7700;font-weight:bold;">except</span> <span style="color: #008000;">IOError</span> <span style="color: #ff7700;font-weight:bold;">as</span> e:
    <span style="color: #ff7700;font-weight:bold;">print</span><span style="color: black;">&#40;</span><span style="color: #483d8b;">&quot;args: &quot;</span><span style="color: #66cc66;">,</span> e.<span style="color: black;">args</span><span style="color: black;">&#41;</span>
    <span style="color: #ff7700;font-weight:bold;">print</span><span style="color: black;">&#40;</span><span style="color: #483d8b;">&quot;errno: &quot;</span><span style="color: #66cc66;">,</span> e.<span style="color: #dc143c;">errno</span><span style="color: black;">&#41;</span>
    <span style="color: #ff7700;font-weight:bold;">print</span><span style="color: black;">&#40;</span><span style="color: #483d8b;">&quot;filename: &quot;</span><span style="color: #66cc66;">,</span> e.<span style="color: black;">filename</span><span style="color: black;">&#41;</span>
    <span style="color: #ff7700;font-weight:bold;">print</span><span style="color: black;">&#40;</span><span style="color: #483d8b;">&quot;strerror: &quot;</span><span style="color: #66cc66;">,</span> e.<span style="color: black;">strerror</span><span style="color: black;">&#41;</span>
    <span style="color: #ff7700;font-weight:bold;">raise</span> MonError<span style="color: black;">&#40;</span><span style="color: #483d8b;">&quot;T'as encore oublie ton classeur Man&quot;</span><span style="color: black;">&#41;</span></pre></td></tr></table></div>

<h2>Erreur à la sortie de l&#8217;interpréteur</h2>
<p>Il est possible d&#8217;attraper n&#8217;importe quelle exception non gérée juste avant qu&#8217;elle fasse crasher le programme, sans avoir à mettre tout son code dans un gros <code>try/except</code>.</p>
<p>Il faut définir une fonction qui accepte 3 arguments que sont la classe de l&#8217;exception, l&#8217;instance de l&#8217;exception et l&#8217;objet traceback :</p>

<div class="wp_syntax"><table><tr><td class="code"><pre class="python" style="font-family:monospace;"><span style="color: #ff7700;font-weight:bold;">def</span> attrapez_les_tous<span style="color: black;">&#40;</span><span style="color: #008000;">type</span><span style="color: #66cc66;">,</span> value<span style="color: #66cc66;">,</span> <span style="color: #dc143c;">traceback</span><span style="color: black;">&#41;</span>:
    <span style="color: #ff7700;font-weight:bold;">print</span><span style="color: black;">&#40;</span><span style="color: #483d8b;">&quot;Pokemon !&quot;</span><span style="color: black;">&#41;</span></pre></td></tr></table></div>

<p>Ensuite il faut l&#8217;attacher au module <code>sys</code> avec le bon nom :</p>

<div class="wp_syntax"><table><tr><td class="code"><pre class="python" style="font-family:monospace;"><span style="color: #ff7700;font-weight:bold;">import</span> <span style="color: #dc143c;">sys</span>
<span style="color: #dc143c;">sys</span>.<span style="color: black;">excepthook</span> <span style="color: #66cc66;">=</span> attrapez_les_tous</pre></td></tr></table></div>

<p>Et hop :</p>

<div class="wp_syntax"><table><tr><td class="code"><pre class="python" style="font-family:monospace;"><span style="color: #ff4500;">1</span> / <span style="color: #ff4500;">0</span>
Pokemon <span style="color: #66cc66;">!</span></pre></td></tr></table></div>

<p>Alors, c&#8217;est certain que si vous faites ça, vous avez intérêt à savoir ce que vous branlez car vous pouver tuer le debuggage. Ou alors créer un super moyen de logger toute erreur sur un process séparé qui vous le présente dans une belle interface Web. Au choix.</p>
<h2>Les exceptions en dehors des erreurs</h2>
<p>Les exceptions sont la pour gérer des cas exceptionnels, et pas forcément une erreur. Par exemple, la boucle <code>for</code> utilise les exceptions pour savoir quand s&#8217;arrêter. En effet elle attend que l&#8217;itérateur qu&#8217;elle utilise lève <code>StopIteration</code> :</p>

<div class="wp_syntax"><table><tr><td class="code"><pre class="python" style="font-family:monospace;"><span style="color: #66cc66;">&gt;&gt;&gt;</span> iterateur <span style="color: #66cc66;">=</span> <span style="color: #008000;">iter</span><span style="color: black;">&#40;</span><span style="color: #008000;">range</span><span style="color: black;">&#40;</span><span style="color: #ff4500;">3</span><span style="color: black;">&#41;</span><span style="color: black;">&#41;</span>
<span style="color: #66cc66;">&gt;&gt;&gt;</span> next<span style="color: black;">&#40;</span>iterateur<span style="color: black;">&#41;</span>
<span style="color: #ff4500;">0</span>
<span style="color: #66cc66;">&gt;&gt;&gt;</span> next<span style="color: black;">&#40;</span>iterateur<span style="color: black;">&#41;</span>
<span style="color: #ff4500;">1</span>
<span style="color: #66cc66;">&gt;&gt;&gt;</span> next<span style="color: black;">&#40;</span>iterateur<span style="color: black;">&#41;</span>
<span style="color: #ff4500;">2</span>
<span style="color: #66cc66;">&gt;&gt;&gt;</span> next<span style="color: black;">&#40;</span>iterateur<span style="color: black;">&#41;</span>
---------------------------------------------------------------------------
<span style="color: #008000;">StopIteration</span>                             Traceback <span style="color: black;">&#40;</span>most recent call last<span style="color: black;">&#41;</span>
<span style="color: #66cc66;">&lt;</span>ipython-input-<span style="color: #ff4500;">11</span>-d90e1e04a685<span style="color: #66cc66;">&gt;</span> <span style="color: #ff7700;font-weight:bold;">in</span> <span style="color: #66cc66;">&lt;</span>module<span style="color: #66cc66;">&gt;</span><span style="color: black;">&#40;</span><span style="color: black;">&#41;</span>
----<span style="color: #66cc66;">&gt;</span> <span style="color: #ff4500;">1</span> next<span style="color: black;">&#40;</span>iterateur<span style="color: black;">&#41;</span>
&nbsp;
<span style="color: #008000;">StopIteration</span>:</pre></td></tr></table></div>

<p>Donc quand on fait ça :</p>

<div class="wp_syntax"><table><tr><td class="code"><pre class="python" style="font-family:monospace;"><span style="color: #ff7700;font-weight:bold;">for</span> x <span style="color: #ff7700;font-weight:bold;">in</span> <span style="color: #008000;">range</span><span style="color: black;">&#40;</span><span style="color: #ff4500;">3</span><span style="color: black;">&#41;</span>:
    <span style="color: #ff7700;font-weight:bold;">print</span><span style="color: black;">&#40;</span>x<span style="color: black;">&#41;</span></pre></td></tr></table></div>

<p>En fait, la boucle fait un truc comme ça :</p>

<div class="wp_syntax"><table><tr><td class="code"><pre class="python" style="font-family:monospace;">iterateur <span style="color: #66cc66;">=</span> <span style="color: #008000;">iter</span><span style="color: black;">&#40;</span><span style="color: #008000;">range</span><span style="color: black;">&#40;</span><span style="color: #ff4500;">3</span><span style="color: black;">&#41;</span><span style="color: black;">&#41;</span>
<span style="color: #ff7700;font-weight:bold;">while</span> <span style="color: #008000;">True</span>:
    <span style="color: #ff7700;font-weight:bold;">try</span>:
        <span style="color: #ff7700;font-weight:bold;">print</span><span style="color: black;">&#40;</span>next<span style="color: black;">&#40;</span>iterateur<span style="color: black;">&#41;</span><span style="color: black;">&#41;</span>
    <span style="color: #ff7700;font-weight:bold;">except</span> <span style="color: #008000;">StopIteration</span>:
        <span style="color: #ff7700;font-weight:bold;">break</span></pre></td></tr></table></div>

<p>Les exceptions en Python, c&#8217;est une grande histoire d&#8217;amour.</p>
]]></content:encoded>
			<wfw:commentRss>http://sametmax.com/gestion-des-erreurs-en-python/feed/</wfw:commentRss>
		<slash:comments>12</slash:comments>
	<enclosure url="http://sametmax.com/wp-content/uploads/2015/01/epic-fail-catch.jpg" length="47973" type="image/jpg" />	</item>
		<item>
		<title>Log post mortem avec Python 8</title>
		<link>http://sametmax.com/log-post-mortem-avec-python/</link>
		<comments>http://sametmax.com/log-post-mortem-avec-python/#comments</comments>
		<pubDate>Thu, 07 Mar 2013 11:25:58 +0000</pubDate>
		<dc:creator><![CDATA[Sam]]></dc:creator>
				<category><![CDATA[Programmation]]></category>
		<category><![CDATA[callback]]></category>
		<category><![CDATA[django]]></category>
		<category><![CDATA[exceptions]]></category>
		<category><![CDATA[python]]></category>

		<guid isPermaLink="false">http://sametmax.com/?p=5260</guid>
		<description><![CDATA[Recette de callback sur un crash de la VM suite à la levée d'une exception.]]></description>
				<content:encoded><![CDATA[<p>Il y a des mois de ça, j&#8217;avais écris <a href="http://sametmax.com/lancer-une-fonction-automatiquement-a-larret-de-python/">un article sur atexit</a>, un module qui permet de lancer une fonction à la fermeture de la VM Python.</p>
<p>Ces fonctions sont appelées même si la VM s&#8217;arrête brutalement, mais on a aucune information sur la raison de l&#8217;arrêt de la machine virtuelle Python. Et elles sont exécutées même si tout s&#8217;est bien passé.</p>
<p>Si vous voulez réagir au plantage de votre programme, et seulement au plantage, tout en ayant en plus des informations sur la nature du foinage :</p>

<div class="wp_syntax"><table><tr><td class="code"><pre class="python" style="font-family:monospace;"><span style="color: #ff7700;font-weight:bold;">import</span> <span style="color: #dc143c;">sys</span>
&nbsp;
<span style="color: #ff7700;font-weight:bold;">def</span> on_crash<span style="color: black;">&#40;</span><span style="color: #008000;">type</span><span style="color: #66cc66;">,</span> value<span style="color: #66cc66;">,</span> tb<span style="color: black;">&#41;</span>:
    <span style="color: #ff7700;font-weight:bold;">print</span> <span style="color: #008000;">type</span>
    <span style="color: #ff7700;font-weight:bold;">print</span> value
    <span style="color: #ff7700;font-weight:bold;">print</span> tb
&nbsp;
<span style="color: #dc143c;">sys</span>.<span style="color: black;">excepthook</span> <span style="color: #66cc66;">=</span> on_crash
&nbsp;
declencher_erreur <span style="color: #66cc66;">=</span> <span style="color: #ff4500;">1</span> + <span style="color: #483d8b;">&quot;1&quot;</span>
&nbsp;
<span style="color: #808080; font-style: italic;">## &lt;type 'exceptions.TypeError'&gt;</span>
<span style="color: #808080; font-style: italic;">## unsupported operand type(s) for +: 'int' and 'str'</span>
<span style="color: #808080; font-style: italic;">## &lt;traceback object at 0x00543AF8&gt;</span></pre></td></tr></table></div>

<p>Ca peut être très intéressant pour débugger un processus détaché comme par exemple votre serveur WSGI qui fait tourner Django qui décide de se petit-suicider : demandez à <code>sys.excepthook</code> de faire un dump de l&#8217;exception dans un fichier log , et vous pourrez voir ce qui a propoqué la crise.</p>
<p>Si vous êtes du genre poli, vous voudrez quand même garder l&#8217;ancien comportement de <code>sys.excepthook</code>, qui est toujours disponible depuis <code>sys.__excepthook__</code> :</p>

<div class="wp_syntax"><table><tr><td class="code"><pre class="python" style="font-family:monospace;"><span style="color: #ff7700;font-weight:bold;">def</span> on_crash<span style="color: black;">&#40;</span><span style="color: #008000;">type</span><span style="color: #66cc66;">,</span> value<span style="color: #66cc66;">,</span> tb<span style="color: black;">&#41;</span>:
    <span style="color: #808080; font-style: italic;"># faire ce que vous voulez ici puis...</span>
    <span style="color: #dc143c;">sys</span>.__excepthook__<span style="color: black;">&#40;</span><span style="color: #008000;">type</span><span style="color: #66cc66;">,</span> value<span style="color: #66cc66;">,</span> tb<span style="color: black;">&#41;</span></pre></td></tr></table></div>

<p>Et dire que vous commenciez à croire que vous saviez presque tout sur Python. Mouarf.</p>
]]></content:encoded>
			<wfw:commentRss>http://sametmax.com/log-post-mortem-avec-python/feed/</wfw:commentRss>
		<slash:comments>8</slash:comments>
	<enclosure url="http://sametmax.com/wp-content/uploads/2013/03/the-walking-dead-3-the-walking-dead-24037474-500-333.jpg" length="54725" type="image/jpg" />	</item>
		<item>
		<title>Pourquoi il faut spécifier l&#8217;exception qu&#8217;on gère 21</title>
		<link>http://sametmax.com/pourquoi-il-faut-specifier-lexception-quon-gere/</link>
		<comments>http://sametmax.com/pourquoi-il-faut-specifier-lexception-quon-gere/#comments</comments>
		<pubDate>Mon, 18 Feb 2013 14:33:01 +0000</pubDate>
		<dc:creator><![CDATA[Sam]]></dc:creator>
				<category><![CDATA[Programmation]]></category>
		<category><![CDATA[exceptions]]></category>
		<category><![CDATA[python]]></category>

		<guid isPermaLink="false">http://sametmax.com/?p=4562</guid>
		<description><![CDATA[On utilise beaucoup la gestion des exceptions en Python, mais le langage permet de faire des trucs apocryphes.]]></description>
				<content:encoded><![CDATA[<p>On utilise beaucoup la gestion des exceptions en Python, mais le langage permet de faire des trucs apocryphes comme:</p>

<div class="wp_syntax"><table><tr><td class="code"><pre class="python" style="font-family:monospace;"><span style="color: #ff7700;font-weight:bold;">try</span>:
    <span style="color: #808080; font-style: italic;"># un code qui va chier des bulles</span>
<span style="color: #ff7700;font-weight:bold;">except</span>: <span style="color: #808080; font-style: italic;"># je les attrappe tous, comme les pokemons</span>
    <span style="color: #ff7700;font-weight:bold;">pass</span></pre></td></tr></table></div>

<p>Ceci n&#8217;est ni plus ni moins que l&#8217;équivalent de <code>@</code> pour silent les erreurs en PHP.</p>
<p>C&#8217;est caca.</p>
<p>Et la raison pour laquelle j&#8217;oppose une aversion si mesurée mais néanmoins portée par des arguments forts, c&#8217;est que c&#8217;est le grand kiff de Max d&#8217;en mettre partout, et que la semaine dernière ça lui a valu un debuggage en prod.</p>
<p>Depuis le dernier déploiement, les logs d&#8217;un des workers d&#8217;un serveur d&#8217;encoding affichaient en boucle :</p>
<blockquote><p>&#8220;Error accessing Database, check your connexion&#8221;</p></blockquote>
<p>Du coup Max a cherché d&#8217;ou venait ce problème de database par les cas usuels : port ? Permission ? Network ? Mise à jour de lib de l&#8217;ORM ? Daemonisation ?</p>
<p>Le snippet qui provoquait ce message était (en simplifié) :</p>

<div class="wp_syntax"><table><tr><td class="code"><pre class="python" style="font-family:monospace;"><span style="color: #ff7700;font-weight:bold;">try</span>:
    new_model <span style="color: #66cc66;">=</span> Model.<span style="color: black;">objects</span>.<span style="color: black;">create</span><span style="color: black;">&#40;</span>params<span style="color: black;">&#41;</span>
<span style="color: #ff7700;font-weight:bold;">except</span>:
    <span style="color: #008000;">self</span>.<span style="color: black;">logger</span>.<span style="color: black;">error</span><span style="color: black;">&#40;</span><span style="color: #483d8b;">&quot;Error accessing Database, check your connexion&quot;</span><span style="color: black;">&#41;</span>
    <span style="color: #808080; font-style: italic;"># + trucs pour gérer le bouzin</span></pre></td></tr></table></div>

<p>La raison du <code>try</code> / <code>except</code> est que parfois, sur ce serveur, et seulement sur ce serveur, la connection avec la DB distante droppait. Bug aléatoire ou de config ? Parfois il est plus simple de ne pas résoudre le problème et laisser le truc foirer de temps en temps, et juste garder un recovery. Ça évite un spécial case pour un seul serveur, et ça évite de migrer le serveur.</p>
<p>Sauf que ce <code>try</code> / <code>except</code> était générique, et il attrapait tout, même les rhumes. </p>
<p>Je lui ai demandé de virer le <code>try</code> / <code>except</code> pour voir ce qui se passait vraiment, et&#8230;</p>

<div class="wp_syntax"><table><tr><td class="code"><pre class="python" style="font-family:monospace;"><span style="color: #ff7700;font-weight:bold;">raise</span> MissingDependency<span style="color: black;">&#40;</span><span style="color: #483d8b;">&quot;The 'solr' backend requires the installation of 'pysolr'. Please refer to the documentation.&quot;</span><span style="color: black;">&#41;</span></pre></td></tr></table></div>

<p>Le truc qui n&#8217;a RIEN à voir avec la base de données. Solr est notre moteur de recherche, et c&#8217;est juste qu&#8217;à la création d&#8217;un objet de modèle Django, un signal est propagé pour mettre à jour l&#8217;index de Solr. Le nouveau déploiement était incomplet : il manquait une lib. La lib était importée à ce moment là, ça levait une exception, qui était attrapée bêtement par le <code>try</code>/<code>except</code> générique.</p>
<p>Bref, le code a été remplacé par</p>

<div class="wp_syntax"><table><tr><td class="code"><pre class="python" style="font-family:monospace;"><span style="color: #ff7700;font-weight:bold;">try</span>:
    new_model <span style="color: #66cc66;">=</span> Model.<span style="color: black;">objects</span>.<span style="color: black;">create</span><span style="color: black;">&#40;</span>params<span style="color: black;">&#41;</span>
<span style="color: #ff7700;font-weight:bold;">except</span> DataBaseError:
    <span style="color: #008000;">self</span>.<span style="color: black;">logger</span>.<span style="color: black;">error</span><span style="color: black;">&#40;</span><span style="color: #483d8b;">&quot;Error accessing Database, check your connexion&quot;</span><span style="color: black;">&#41;</span>
    <span style="color: #808080; font-style: italic;"># + trucs pour gérer le bouzin</span></pre></td></tr></table></div>

<p>Si on l&#8217;avait eu depuis le début, l&#8217;erreur aurait été apparente tout de suite et rapidement corrigée.</p>
<p>Ne faites JAMAIS de <code>try</code>/<code>except</code> sans préciser l&#8217;exception que vous allez gérer. La seule raison pour laquelle cette feature existe, c&#8217;est qu&#8217;elle est pratique à utiliser dans un shell. En fait, même <code>DataBaseError</code> est trop générique comme Exception si vous voulez vraiment un controle fin.</p>
<p>Notez également qu&#8217;un <code>except: pass</code> est rarement une bonne idée. Au moins prenez le temps de logger ce qui se passe dans la clause <code>except</code>. Un bon log, c&#8217;est le début du bonheur.</p>
]]></content:encoded>
			<wfw:commentRss>http://sametmax.com/pourquoi-il-faut-specifier-lexception-quon-gere/feed/</wfw:commentRss>
		<slash:comments>21</slash:comments>
	<enclosure url="http://sametmax.com/wp-content/uploads/2013/02/2205118143_561cdd1947.jpg" length="85039" type="image/jpg" />	</item>
		<item>
		<title>Comment recruter un développeur Python 7</title>
		<link>http://sametmax.com/comment-recruter-un-developpeur-python/</link>
		<comments>http://sametmax.com/comment-recruter-un-developpeur-python/#comments</comments>
		<pubDate>Sat, 27 Oct 2012 14:28:54 +0000</pubDate>
		<dc:creator><![CDATA[Sam]]></dc:creator>
				<category><![CDATA[Programmation]]></category>
		<category><![CDATA[exceptions]]></category>
		<category><![CDATA[humour]]></category>
		<category><![CDATA[python]]></category>

		<guid isPermaLink="false">http://sametmax.com/?p=2759</guid>
		<description><![CDATA[Bonjour M. Gentil. Vous venez pour la position de stagiaire ingénieur senior en periode d'essai sur 3 ans ?

C'est bien, c'est bien.

Nous avons des perspectives de progression fascinantes dans notre SS3I au carré.

J'ai juste quelques tests à vous faire passer. Trois fois rien. Simple formalité administrative. ]]></description>
				<content:encoded><![CDATA[<p>Bonjour M. Gentil. Vous venez pour la position de stagiaire ingénieur senior en periode d&#8217;essai sur 3 ans ?</p>
<p>C&#8217;est bien, c&#8217;est bien.</p>
<p>Nous avons des perspectives de progression fascinantes dans notre SS3I au carré.</p>
<p>J&#8217;ai juste quelques tests à vous faire passer. Trois fois rien. Simple formalité administrative. Vous comprenez, on ne peut pas embaucher pas n&#8217;importe quel Bac + 5 et le payer SMIC, comme ça sur un coup de tête.</p>
<p>Ce n&#8217;est pas contre vous, non.</p>
<p>Pourriez-vous me dire ce qu&#8217;affiche ce snippet ? (mouahahahahaha, rire diabolique intérieur)</p>

<div class="wp_syntax"><table><tr><td class="code"><pre class="python" style="font-family:monospace;"><span style="color: #ff7700;font-weight:bold;">def</span> <span style="color: #dc143c;">test</span><span style="color: black;">&#40;</span><span style="color: black;">&#41;</span>:
&nbsp;
    <span style="color: #ff7700;font-weight:bold;">try</span>:
        <span style="color: #ff7700;font-weight:bold;">return</span> <span style="color: #ff4500;">1</span> + <span style="color: #483d8b;">&quot;1&quot;</span>
    <span style="color: #ff7700;font-weight:bold;">except</span> <span style="color: #008000;">TypeError</span>:
        <span style="color: #ff7700;font-weight:bold;">return</span> <span style="color: #483d8b;">&quot;exception&quot;</span>
    <span style="color: #ff7700;font-weight:bold;">finally</span>:
        <span style="color: #ff7700;font-weight:bold;">return</span> <span style="color: #483d8b;">&quot;finally&quot;</span>
&nbsp;
<span style="color: #ff7700;font-weight:bold;">print</span> <span style="color: #dc143c;">test</span><span style="color: black;">&#40;</span><span style="color: black;">&#41;</span></pre></td></tr></table></div>

<p>Et celui-là, il affiche la stack trace ou pas ?</p>

<div class="wp_syntax"><table><tr><td class="code"><pre class="python" style="font-family:monospace;"><span style="color: #ff7700;font-weight:bold;">def</span> <span style="color: #dc143c;">test</span><span style="color: black;">&#40;</span><span style="color: black;">&#41;</span>:
&nbsp;
    <span style="color: #ff7700;font-weight:bold;">try</span>:
        <span style="color: #ff7700;font-weight:bold;">print</span> <span style="color: #ff4500;">1</span> + <span style="color: #483d8b;">&quot;1&quot;</span>
    <span style="color: #ff7700;font-weight:bold;">except</span> <span style="color: #008000;">TypeError</span>:
        <span style="color: #ff7700;font-weight:bold;">raise</span> <span style="color: #008000;">ValueError</span><span style="color: black;">&#40;</span><span style="color: #483d8b;">'Test'</span><span style="color: black;">&#41;</span>
    <span style="color: #ff7700;font-weight:bold;">finally</span>:
        <span style="color: #ff7700;font-weight:bold;">return</span> <span style="color: #483d8b;">&quot;finally&quot;</span>
&nbsp;
<span style="color: #ff7700;font-weight:bold;">print</span> <span style="color: #dc143c;">test</span><span style="color: black;">&#40;</span><span style="color: black;">&#41;</span></pre></td></tr></table></div>

<p>Bon ok, mais si on a un générateur alors ?</p>

<div class="wp_syntax"><table><tr><td class="code"><pre class="python" style="font-family:monospace;"><span style="color: #ff7700;font-weight:bold;">def</span> <span style="color: #dc143c;">test</span><span style="color: black;">&#40;</span><span style="color: black;">&#41;</span>:
&nbsp;
    <span style="color: #ff7700;font-weight:bold;">try</span>:
        <span style="color: #ff7700;font-weight:bold;">yield</span> <span style="color: #ff4500;">1</span> + <span style="color: #483d8b;">&quot;1&quot;</span>
    <span style="color: #ff7700;font-weight:bold;">except</span> <span style="color: #008000;">TypeError</span>:
        <span style="color: #ff7700;font-weight:bold;">yield</span> <span style="color: #483d8b;">'typerror'</span>
        <span style="color: #ff7700;font-weight:bold;">return</span>
    <span style="color: #ff7700;font-weight:bold;">finally</span>:
        <span style="color: #ff7700;font-weight:bold;">yield</span> <span style="color: #483d8b;">&quot;finally&quot;</span>
        <span style="color: #ff7700;font-weight:bold;">return</span>
&nbsp;
    <span style="color: #ff7700;font-weight:bold;">yield</span> <span style="color: #483d8b;">&quot;Out&quot;</span>
&nbsp;
<span style="color: #ff7700;font-weight:bold;">for</span> value <span style="color: #ff7700;font-weight:bold;">in</span> <span style="color: #dc143c;">test</span><span style="color: black;">&#40;</span><span style="color: black;">&#41;</span>:
    <span style="color: #ff7700;font-weight:bold;">print</span> value</pre></td></tr></table></div>

<p>Quelle exception sera catchée, dans le bout de code suivant ? (tapoter son style de manière énervante sur la table)</p>

<div class="wp_syntax"><table><tr><td class="code"><pre class="python" style="font-family:monospace;"><span style="color: #ff7700;font-weight:bold;">def</span> <span style="color: #dc143c;">test</span><span style="color: black;">&#40;</span><span style="color: black;">&#41;</span>:
&nbsp;
    <span style="color: #ff7700;font-weight:bold;">try</span>:
        <span style="color: #ff7700;font-weight:bold;">assert</span> <span style="color: #ff4500;">1</span> + <span style="color: #483d8b;">&quot;1&quot;</span>
    <span style="color: #ff7700;font-weight:bold;">except</span> <span style="color: #008000;">AssertionError</span>:
        <span style="color: #ff7700;font-weight:bold;">print</span> <span style="color: #483d8b;">&quot;assertionerror&quot;</span>
    <span style="color: #ff7700;font-weight:bold;">except</span> <span style="color: #008000;">Exception</span>:
        <span style="color: #ff7700;font-weight:bold;">print</span> <span style="color: #483d8b;">&quot;exception&quot;</span>
    <span style="color: #ff7700;font-weight:bold;">except</span> <span style="color: #008000;">TypeError</span>:
        <span style="color: #ff7700;font-weight:bold;">print</span> <span style="color: #483d8b;">&quot;typerror&quot;</span>
&nbsp;
<span style="color: #dc143c;">test</span><span style="color: black;">&#40;</span><span style="color: black;">&#41;</span></pre></td></tr></table></div>

<p>Je vois. Donc finally est exécuté dans tous les cas ?</p>

<div class="wp_syntax"><table><tr><td class="code"><pre class="python" style="font-family:monospace;"><span style="color: #ff7700;font-weight:bold;">def</span> <span style="color: #dc143c;">test</span><span style="color: black;">&#40;</span><span style="color: black;">&#41;</span>:
&nbsp;
    <span style="color: #ff7700;font-weight:bold;">def</span> foo<span style="color: black;">&#40;</span><span style="color: black;">&#41;</span>:
        foo<span style="color: black;">&#40;</span><span style="color: black;">&#41;</span>
&nbsp;
    <span style="color: #ff7700;font-weight:bold;">try</span>:
        foo<span style="color: black;">&#40;</span><span style="color: black;">&#41;</span>
    <span style="color: #ff7700;font-weight:bold;">except</span> <span style="color: #008000;">RuntimeError</span>:
        <span style="color: #ff7700;font-weight:bold;">print</span> <span style="color: #483d8b;">'runtimeerror'</span>
    <span style="color: #ff7700;font-weight:bold;">finally</span>:
        <span style="color: #ff7700;font-weight:bold;">print</span> <span style="color: #483d8b;">'finally'</span>
&nbsp;
<span style="color: #dc143c;">test</span><span style="color: black;">&#40;</span><span style="color: black;">&#41;</span></pre></td></tr></table></div>

<p>Vous êtes sûr ? VRAIMENT dans tous les cas ? (regard appuyé bien stressant, travaillé durant un poste de manager chez Quick)</p>

<div class="wp_syntax"><table><tr><td class="code"><pre class="python" style="font-family:monospace;"><span style="color: #ff7700;font-weight:bold;">def</span> <span style="color: #dc143c;">test</span><span style="color: black;">&#40;</span><span style="color: black;">&#41;</span>:
&nbsp;
    <span style="color: #ff7700;font-weight:bold;">try</span>:
        PRINTEUH <span style="color: #66cc66;">!</span>
    <span style="color: #ff7700;font-weight:bold;">except</span> <span style="color: #008000;">SyntaxError</span>:
        <span style="color: #ff7700;font-weight:bold;">print</span> <span style="color: #483d8b;">'syntaxerror'</span>
    <span style="color: #ff7700;font-weight:bold;">finally</span>:
        <span style="color: #ff7700;font-weight:bold;">print</span> <span style="color: #483d8b;">'finally'</span>
&nbsp;
<span style="color: #dc143c;">test</span><span style="color: black;">&#40;</span><span style="color: black;">&#41;</span></pre></td></tr></table></div>

<p>VRAIMENT, VRAIMENT, dans tous les cas ? Mais alors, VRAIMENT ? (prendre la voix d&#8217;Alain Chabat, parce qu&#8217;arrivé à ce stade là c&#8217;est juste plus rigolo)</p>

<div class="wp_syntax"><table><tr><td class="code"><pre class="python" style="font-family:monospace;"><span style="color: #ff7700;font-weight:bold;">def</span> <span style="color: #dc143c;">test</span><span style="color: black;">&#40;</span><span style="color: black;">&#41;</span>:
&nbsp;
    <span style="color: #ff7700;font-weight:bold;">def</span> foo<span style="color: black;">&#40;</span>bar<span style="color: black;">&#41;</span>:
        <span style="color: #ff7700;font-weight:bold;">print</span> bar
        <span style="color: #ff7700;font-weight:bold;">return</span> foo<span style="color: #66cc66;">,</span> bar
&nbsp;
    bar <span style="color: #66cc66;">=</span> <span style="color: #483d8b;">'bar'</span>
&nbsp;
    <span style="color: #ff7700;font-weight:bold;">try</span>:
        <span style="color: #ff7700;font-weight:bold;">while</span> <span style="color: #008000;">True</span>:
            foo<span style="color: #66cc66;">,</span> bar <span style="color: #66cc66;">=</span> foo<span style="color: black;">&#40;</span>bar<span style="color: black;">&#41;</span>
&nbsp;
    <span style="color: #ff7700;font-weight:bold;">except</span> <span style="color: #008000;">RuntimeError</span>:
        <span style="color: #ff7700;font-weight:bold;">print</span> <span style="color: #483d8b;">'runtimeerror'</span>
    <span style="color: #ff7700;font-weight:bold;">finally</span>:
        <span style="color: #ff7700;font-weight:bold;">print</span> <span style="color: #483d8b;">'finally'</span>
&nbsp;
<span style="color: #dc143c;">test</span><span style="color: black;">&#40;</span><span style="color: black;">&#41;</span></pre></td></tr></table></div>

<p>Ne vous inquiétez pas, nous ne vous JUgeons paaaaaaas. Détendez-vous. Allez. Une petite dernière. Je vous aide. <code>sys.exit(1)</code> retourne le code 1.</p>

<div class="wp_syntax"><table><tr><td class="code"><pre class="python" style="font-family:monospace;"><span style="color: #ff7700;font-weight:bold;">import</span> <span style="color: #dc143c;">sys</span>
&nbsp;
<span style="color: #ff7700;font-weight:bold;">def</span> <span style="color: #dc143c;">test</span><span style="color: black;">&#40;</span>l<span style="color: #66cc66;">=</span><span style="color: black;">&#91;</span><span style="color: black;">&#93;</span><span style="color: black;">&#41;</span>:
&nbsp;
    <span style="color: #ff7700;font-weight:bold;">print</span> l
    <span style="color: #ff7700;font-weight:bold;">try</span>:
        <span style="color: #ff7700;font-weight:bold;">return</span> <span style="color: #ff4500;">1</span> + <span style="color: #483d8b;">&quot;1&quot;</span>
    <span style="color: #ff7700;font-weight:bold;">except</span> <span style="color: #008000;">TypeError</span>:
        <span style="color: #ff7700;font-weight:bold;">return</span> l <span style="color: #ff7700;font-weight:bold;">or</span> l.<span style="color: black;">append</span><span style="color: black;">&#40;</span><span style="color: #dc143c;">sys</span>.<span style="color: black;">exit</span><span style="color: black;">&#40;</span><span style="color: #ff4500;">1</span><span style="color: black;">&#41;</span><span style="color: black;">&#41;</span>
    <span style="color: #ff7700;font-weight:bold;">finally</span>:
        <span style="color: #ff7700;font-weight:bold;">print</span> <span style="color: #483d8b;">&quot;Je sais plus là, sérieux&quot;</span>
        <span style="color: #ff7700;font-weight:bold;">if</span> <span style="color: #008000;">len</span><span style="color: black;">&#40;</span>l<span style="color: black;">&#41;</span> <span style="color: #66cc66;">&lt;</span> <span style="color: #ff4500;">1</span>:
            <span style="color: #dc143c;">test</span><span style="color: black;">&#40;</span><span style="color: black;">&#41;</span>
&nbsp;
<span style="color: #dc143c;">test</span><span style="color: black;">&#40;</span><span style="color: black;">&#41;</span>
<span style="color: #dc143c;">test</span><span style="color: black;">&#40;</span><span style="color: black;">&#41;</span></pre></td></tr></table></div>

<p>Bon, je sens que vous êtes fatigué. On va arrêter là peut être. Vous êtes le genre à vous en tenir là, hein ? Non. Très bien, très bien. Voici un code sur lequel on planche en interne depuis une semaine pour comprendre combien de fois il affiche finally, mais on s&#8217;est dit que vous pourriez le résoudre gratuitement pour nous:</p>

<div class="wp_syntax"><table><tr><td class="code"><pre class="python" style="font-family:monospace;"><span style="color: #ff7700;font-weight:bold;">import</span> <span style="color: #dc143c;">sys</span>
<span style="color: #ff7700;font-weight:bold;">import</span> <span style="color: #dc143c;">time</span>
&nbsp;
<span style="color: #ff7700;font-weight:bold;">import</span> multiprocessing
<span style="color: #ff7700;font-weight:bold;">from</span> <span style="color: #dc143c;">Queue</span> <span style="color: #ff7700;font-weight:bold;">import</span> Empty
&nbsp;
in_queue <span style="color: #66cc66;">=</span> multiprocessing.<span style="color: #dc143c;">Queue</span><span style="color: black;">&#40;</span><span style="color: black;">&#41;</span>
out_queue <span style="color: #66cc66;">=</span> multiprocessing.<span style="color: #dc143c;">Queue</span><span style="color: black;">&#40;</span><span style="color: black;">&#41;</span>
&nbsp;
<span style="color: #ff7700;font-weight:bold;">def</span> worker<span style="color: black;">&#40;</span><span style="color: black;">&#41;</span>:
&nbsp;
    <span style="color: #ff7700;font-weight:bold;">while</span> <span style="color: #008000;">True</span>:
&nbsp;
        <span style="color: #ff7700;font-weight:bold;">try</span>:
            res <span style="color: #66cc66;">=</span> in_queue.<span style="color: black;">get</span><span style="color: black;">&#40;</span>timeout<span style="color: #66cc66;">=</span><span style="color: #ff4500;">0.1</span><span style="color: black;">&#41;</span>
            <span style="color: #ff7700;font-weight:bold;">if</span> res <span style="color: #66cc66;">==</span> <span style="color: #483d8b;">'stop'</span>:
                <span style="color: #dc143c;">sys</span>.<span style="color: black;">exit</span><span style="color: black;">&#40;</span><span style="color: #ff4500;">1</span><span style="color: black;">&#41;</span>
            <span style="color: #ff7700;font-weight:bold;">print</span> res
        <span style="color: #ff7700;font-weight:bold;">except</span> <span style="color: black;">&#40;</span>Empty<span style="color: #66cc66;">,</span> multiprocessing.<span style="color: black;">TimeoutError</span><span style="color: black;">&#41;</span>:
            <span style="color: #ff7700;font-weight:bold;">pass</span>
        <span style="color: #ff7700;font-weight:bold;">finally</span>:
            <span style="color: #ff7700;font-weight:bold;">print</span> <span style="color: #483d8b;">'finally'</span>
&nbsp;
&nbsp;
process <span style="color: #66cc66;">=</span> multiprocessing.<span style="color: black;">Process</span><span style="color: black;">&#40;</span>target<span style="color: #66cc66;">=</span>worker<span style="color: black;">&#41;</span>
process.<span style="color: black;">start</span><span style="color: black;">&#40;</span><span style="color: black;">&#41;</span>
&nbsp;
&nbsp;
in_queue.<span style="color: black;">put</span><span style="color: black;">&#40;</span><span style="color: #483d8b;">'test'</span><span style="color: black;">&#41;</span>
in_queue.<span style="color: black;">put</span><span style="color: black;">&#40;</span><span style="color: #483d8b;">'stop'</span><span style="color: black;">&#41;</span>
&nbsp;
<span style="color: #ff7700;font-weight:bold;">print</span> <span style="color: #483d8b;">'afterstop'</span>
&nbsp;
<span style="color: #dc143c;">time</span>.<span style="color: black;">sleep</span><span style="color: black;">&#40;</span><span style="color: #ff4500;">1</span><span style="color: black;">&#41;</span>
&nbsp;
<span style="color: #ff7700;font-weight:bold;">print</span> <span style="color: #483d8b;">'done'</span></pre></td></tr></table></div>

<p>Merci, ce sera tout.</p>
<p>Bon WE, monsieur Gentil.</p>
<p>Nous gardons votre CV. On vous recontactera.</p>
]]></content:encoded>
			<wfw:commentRss>http://sametmax.com/comment-recruter-un-developpeur-python/feed/</wfw:commentRss>
		<slash:comments>7</slash:comments>
	<enclosure url="http://sametmax.com/wp-content/uploads/2012/10/2004_12_31_millions.jpg" length="34256" type="image/jpg" />	</item>
	</channel>
</rss>
